
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.3, mkdocs-material-8.2.1">
    
    
      
        <title>Exploratory Analysis - Bulk RNAseq data analysis - an introductory course</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.e8d9bf0c.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.e6a45f82.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../extra.css">
    
    <script>__md_scope=new URL("..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#exploratory-analysis" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Bulk RNAseq data analysis - an introductory course" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis - an introductory course" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bulk RNAseq data analysis - an introductory course
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Exploratory Analysis
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Bulk RNAseq data analysis - an introductory course" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis - an introductory course" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    Bulk RNAseq data analysis - an introductory course
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Course Introduction
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../preprocessing/" class="md-nav__link">
        Preprocessing
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Exploratory Analysis
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Exploratory Analysis
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#data-transformations" class="md-nav__link">
    Data transformations
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#samples-comparisons" class="md-nav__link">
    Samples comparisons
  </a>
  
    <nav class="md-nav" aria-label="Samples comparisons">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#heatmap-of-the-sample-to-sample-distances" class="md-nav__link">
    Heatmap of the sample-to-sample distances
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#principal-component-plot-of-the-samples" class="md-nav__link">
    Principal component plot of the samples
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#glimma-plots-interactive-visualizations-of-the-deseq-results-using" class="md-nav__link">
    # Glimma plots Interactive visualizations of the DESeq results using
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plot-counts" class="md-nav__link">
    Plot counts
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../DEA/" class="md-nav__link">
        Differential Expression
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../contrast_design/" class="md-nav__link">
        Complex designs
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../functional_analysis/" class="md-nav__link">
        Functional Analysis
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#data-transformations" class="md-nav__link">
    Data transformations
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#samples-comparisons" class="md-nav__link">
    Samples comparisons
  </a>
  
    <nav class="md-nav" aria-label="Samples comparisons">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#heatmap-of-the-sample-to-sample-distances" class="md-nav__link">
    Heatmap of the sample-to-sample distances
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#principal-component-plot-of-the-samples" class="md-nav__link">
    Principal component plot of the samples
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#glimma-plots-interactive-visualizations-of-the-deseq-results-using" class="md-nav__link">
    # Glimma plots Interactive visualizations of the DESeq results using
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plot-counts" class="md-nav__link">
    Plot counts
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                


<h1 id="exploratory-analysis">Exploratory analysis</h1>
<p>After running the <code>DESeq()</code> function, we can start our RNAseq analysis
by exploring the relationships between our samples. In principle, all
replicates for a specific condition should be similar to each other
after normalization. Should this not be the case, one might need to
remove outlier replicates. This is one of the reasons why RNAseq
experiments need at least 3 replicates per condition.</p>
<h2 id="data-transformations">Data transformations</h2>
<p>For visualization and clustering (exploratory analyses) – it might be
useful to work with transformed versions of the count data. There are
two main methods used for this purpose: variance stabilizing
transformations (<strong>VST</strong>), and the regularized logarithm or <strong>rlog</strong>.
Both transformations produce transformed data on the log2 scale which
has been normalized with respect to library size or other normalization
factors.</p>
<pre><code class="language-r">vsd &lt;- vst(dds) #Variance Stabilizing Transformation, vst is faster with larger number of samples
rld &lt;- rlog(dds) #Regularized 
</code></pre>
<h2 id="samples-comparisons">Samples comparisons</h2>
<p>We can do some sanity checks of the samples. We can see how they
correlate to each other in a heatmap or see their Principal Components
with a PCA plot.</p>
<h3 id="heatmap-of-the-sample-to-sample-distances">Heatmap of the sample-to-sample distances</h3>
<p>This plot shows how far away are each sample from each other. The darker
the blue, the closer they are.</p>
<pre><code class="language-r">sampleDists &lt;- dist(t(assay(vsd)))
sampleDistMatrix &lt;- as.matrix(sampleDists)
rownames(sampleDistMatrix) &lt;- paste(vsd$condition, vsd$type, sep=&quot;-&quot;)
colnames(sampleDistMatrix) &lt;- NULL
colors &lt;- colorRampPalette( rev(brewer.pal(9, &quot;Blues&quot;)) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
</code></pre>
<p><img src="../exploratory_analysis.Rmd/sample_distances_heatmap-1.png" style="display: block; margin: auto;" /></p>
<h3 id="principal-component-plot-of-the-samples">Principal component plot of the samples</h3>
<p>PCA plot using the first two components</p>
<pre><code class="language-r">pcaData &lt;- plotPCA(vsd, intgroup=c(&quot;condition&quot;, &quot;type&quot;), returnData=TRUE)
percentVar &lt;- round(100 * attr(pcaData, &quot;percentVar&quot;))
ggplot(pcaData, aes(PC1, PC2, color=condition, shape=type)) +
  geom_point(size=3) +
  xlab(paste0(&quot;PC1: &quot;,percentVar[1],&quot;% variance&quot;)) +
  ylab(paste0(&quot;PC2: &quot;,percentVar[2],&quot;% variance&quot;)) + 
  coord_fixed() + theme_bw()
</code></pre>
<p><img src="../exploratory_analysis.Rmd/PCA_plot-1.png" style="display: block; margin: auto;" /></p>
<h2 id="glimma-plots-interactive-visualizations-of-the-deseq-results-using"># Glimma plots Interactive visualizations of the DESeq results using</h2>
<p>the <strong>Glimma</strong> package, which provides excellent options for MA, Volcano
and dimensionality reduction (like a PCA) plots. Use the <em>groups</em>
argument to provide the condition or factor of your experiment.
Unfortunately, the interactive plots created here are not fully
compatible when knitting this document. Feel free to explore them using
this <a href="../mds-plot.html">link</a>!</p>
<pre><code class="language-r">glimma_plot &lt;- glimmaMDS(dds, groups = dds$condition)
</code></pre>
<p>Save your interactive plots using the <strong>htmlwidgets</strong> package,
specifically, the <code>saveWidget()</code></p>
<pre><code class="language-r">htmlwidgets::saveWidget(glimma_plot, &quot;mds-plot.html&quot;)
</code></pre>
<h2 id="plot-counts">Plot counts</h2>
<p>It can be useful to examine the counts of reads for a single gene across
the conditions. A simple function for making this plot is plotCounts,
which normalizes counts by the estimated size factors (or normalization
factors if these were used) and adds a pseudocount of 1/2 to allow for
log scale plotting. The counts are grouped by the variables in
<em>intgroup</em> argument, where more than one variable can be specified. You
can select the gene to plot by its name or by numeric index.</p>
<pre><code class="language-r">d &lt;- plotCounts(dds, gene= 1, intgroup=&quot;condition&quot;, 
                returnData=TRUE)
ggplot(d, aes(x=condition, y=count, color = condition)) + 
  geom_point(position=position_jitter(w=0.1,h=0)) + 
  scale_y_log10(breaks=c(25,100,400)) + theme_bw()
</code></pre>
<p><img src="../exploratory_analysis.Rmd/plot_counts_example-1.png" style="display: block; margin: auto;" /></p>
<h1 id="session-info">Session info</h1>
<p>Finally, we create a <code>session_info()</code> table that will allow anyone to
check what versions of R and packages are we using for reproducibility
purposes.</p>
<pre><code class="language-r">devtools::session_info()
</code></pre>
<pre><code>## ─ Session info ───────────────────────────────────────────────────────────────
##  setting  value
##  version  R version 4.1.0 (2021-05-18)
##  os       macOS Big Sur 10.16
##  system   x86_64, darwin17.0
##  ui       X11
##  language (EN)
##  collate  en_US.UTF-8
##  ctype    en_US.UTF-8
##  tz       Europe/Copenhagen
##  date     2022-02-25
##  pandoc   2.14.0.1 @ /usr/local/bin/ (via rmarkdown)
## 
## ─ Packages ───────────────────────────────────────────────────────────────────
##  package              * version  date (UTC) lib source
##  annotate               1.72.0   2021-10-26 [1] Bioconductor
##  AnnotationDbi          1.56.2   2021-11-09 [1] Bioconductor
##  assertthat             0.2.1    2019-03-21 [1] CRAN (R 4.1.0)
##  backports              1.4.1    2021-12-13 [1] CRAN (R 4.1.0)
##  Biobase              * 2.54.0   2021-10-26 [1] Bioconductor
##  BiocGenerics         * 0.40.0   2021-10-26 [1] Bioconductor
##  BiocParallel           1.28.3   2021-12-09 [1] Bioconductor
##  Biostrings             2.62.0   2021-10-26 [1] Bioconductor
##  bit                    4.0.4    2020-08-04 [1] CRAN (R 4.1.0)
##  bit64                  4.0.5    2020-08-30 [1] CRAN (R 4.1.0)
##  bitops                 1.0-7    2021-04-24 [1] CRAN (R 4.1.0)
##  blob                   1.2.2    2021-07-23 [1] CRAN (R 4.1.0)
##  broom                  0.7.10   2021-10-31 [1] CRAN (R 4.1.0)
##  cachem                 1.0.6    2021-08-19 [1] CRAN (R 4.1.0)
##  callr                  3.7.0    2021-04-20 [1] CRAN (R 4.1.0)
##  cellranger             1.1.0    2016-07-27 [1] CRAN (R 4.1.0)
##  cli                    3.1.0    2021-10-27 [1] CRAN (R 4.1.0)
##  colorspace             2.0-2    2021-06-24 [1] CRAN (R 4.1.0)
##  crayon                 1.4.2    2021-10-29 [1] CRAN (R 4.1.0)
##  data.table             1.14.2   2021-09-27 [1] CRAN (R 4.1.0)
##  DBI                    1.1.1    2021-01-15 [1] CRAN (R 4.1.0)
##  dbplyr                 2.1.1    2021-04-06 [1] CRAN (R 4.1.0)
##  DelayedArray           0.20.0   2021-10-26 [1] Bioconductor
##  desc                   1.4.0    2021-09-28 [1] CRAN (R 4.1.0)
##  DESeq2               * 1.34.0   2021-10-26 [1] Bioconductor
##  devtools               2.4.3    2021-11-30 [1] CRAN (R 4.1.0)
##  digest                 0.6.29   2021-12-01 [1] CRAN (R 4.1.0)
##  dplyr                * 1.0.7    2021-06-18 [1] CRAN (R 4.1.0)
##  edgeR                  3.36.0   2021-10-26 [1] Bioconductor
##  ellipsis               0.3.2    2021-04-29 [1] CRAN (R 4.1.0)
##  evaluate               0.14     2019-05-28 [1] CRAN (R 4.1.0)
##  fansi                  0.5.0    2021-05-25 [1] CRAN (R 4.1.0)
##  farver                 2.1.0    2021-02-28 [1] CRAN (R 4.1.0)
##  fastmap                1.1.0    2021-01-25 [1] CRAN (R 4.1.0)
##  forcats              * 0.5.1    2021-01-27 [1] CRAN (R 4.1.0)
##  fs                     1.5.2    2021-12-08 [1] CRAN (R 4.1.0)
##  genefilter             1.76.0   2021-10-26 [1] Bioconductor
##  geneplotter            1.72.0   2021-10-26 [1] Bioconductor
##  generics               0.1.1    2021-10-25 [1] CRAN (R 4.1.0)
##  GenomeInfoDb         * 1.30.0   2021-10-26 [1] Bioconductor
##  GenomeInfoDbData       1.2.7    2021-11-16 [1] Bioconductor
##  GenomicRanges        * 1.46.1   2021-11-18 [1] Bioconductor
##  ggplot2              * 3.3.5    2021-06-25 [1] CRAN (R 4.1.0)
##  Glimma               * 2.4.0    2021-10-26 [1] Bioconductor
##  glue                   1.5.1    2021-11-30 [1] CRAN (R 4.1.0)
##  gprofiler2           * 0.2.1    2021-08-23 [1] CRAN (R 4.1.0)
##  gtable                 0.3.0    2019-03-25 [1] CRAN (R 4.1.0)
##  haven                  2.4.3    2021-08-04 [1] CRAN (R 4.1.0)
##  highr                  0.9      2021-04-16 [1] CRAN (R 4.1.0)
##  hms                    1.1.1    2021-09-26 [1] CRAN (R 4.1.0)
##  htmltools              0.5.2    2021-08-25 [1] CRAN (R 4.1.0)
##  htmlwidgets            1.5.4    2021-09-08 [1] CRAN (R 4.1.0)
##  httr                   1.4.2    2020-07-20 [1] CRAN (R 4.1.0)
##  IRanges              * 2.28.0   2021-10-26 [1] Bioconductor
##  jsonlite               1.7.2    2020-12-09 [1] CRAN (R 4.1.0)
##  KEGGREST               1.34.0   2021-10-26 [1] Bioconductor
##  knitr                  1.36     2021-09-29 [1] CRAN (R 4.1.0)
##  labeling               0.4.2    2020-10-20 [1] CRAN (R 4.1.0)
##  lattice                0.20-45  2021-09-22 [1] CRAN (R 4.1.0)
##  lazyeval               0.2.2    2019-03-15 [1] CRAN (R 4.1.0)
##  lifecycle              1.0.1    2021-09-24 [1] CRAN (R 4.1.0)
##  limma                  3.50.0   2021-10-26 [1] Bioconductor
##  locfit                 1.5-9.4  2020-03-25 [1] CRAN (R 4.1.0)
##  lubridate              1.8.0    2021-10-07 [1] CRAN (R 4.1.0)
##  magrittr               2.0.1    2020-11-17 [1] CRAN (R 4.1.0)
##  Matrix                 1.4-0    2021-12-08 [1] CRAN (R 4.1.0)
##  MatrixGenerics       * 1.6.0    2021-10-26 [1] Bioconductor
##  matrixStats          * 0.61.0   2021-09-17 [1] CRAN (R 4.1.0)
##  memoise                2.0.1    2021-11-26 [1] CRAN (R 4.1.0)
##  modelr                 0.1.8    2020-05-19 [1] CRAN (R 4.1.0)
##  munsell                0.5.0    2018-06-12 [1] CRAN (R 4.1.0)
##  pheatmap             * 1.0.12   2019-01-04 [1] CRAN (R 4.1.0)
##  pillar                 1.6.4    2021-10-18 [1] CRAN (R 4.1.0)
##  pkgbuild               1.3.0    2021-12-09 [1] CRAN (R 4.1.0)
##  pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 4.1.0)
##  pkgload                1.2.4    2021-11-30 [1] CRAN (R 4.1.0)
##  plotly                 4.10.0   2021-10-09 [1] CRAN (R 4.1.0)
##  png                    0.1-7    2013-12-03 [1] CRAN (R 4.1.0)
##  prettyunits            1.1.1    2020-01-24 [1] CRAN (R 4.1.0)
##  processx               3.5.2    2021-04-30 [1] CRAN (R 4.1.0)
##  ps                     1.6.0    2021-02-28 [1] CRAN (R 4.1.0)
##  purrr                * 0.3.4    2020-04-17 [1] CRAN (R 4.1.0)
##  R6                     2.5.1    2021-08-19 [1] CRAN (R 4.1.0)
##  RColorBrewer         * 1.1-2    2014-12-07 [1] CRAN (R 4.1.0)
##  Rcpp                   1.0.7    2021-07-07 [1] CRAN (R 4.1.0)
##  RCurl                  1.98-1.5 2021-09-17 [1] CRAN (R 4.1.0)
##  readr                * 2.1.1    2021-11-30 [1] CRAN (R 4.1.0)
##  readxl                 1.3.1    2019-03-13 [1] CRAN (R 4.1.0)
##  remotes                2.4.2    2021-11-30 [1] CRAN (R 4.1.0)
##  reprex                 2.0.1    2021-08-05 [1] CRAN (R 4.1.0)
##  rlang                  0.4.12   2021-10-18 [1] CRAN (R 4.1.0)
##  rmarkdown              2.11     2021-09-14 [1] CRAN (R 4.1.0)
##  rprojroot              2.0.2    2020-11-15 [1] CRAN (R 4.1.0)
##  RSQLite                2.2.9    2021-12-06 [1] CRAN (R 4.1.0)
##  rstudioapi             0.13     2020-11-12 [1] CRAN (R 4.1.0)
##  rvest                  1.0.2    2021-10-16 [1] CRAN (R 4.1.0)
##  S4Vectors            * 0.32.3   2021-11-21 [1] Bioconductor
##  scales                 1.1.1    2020-05-11 [1] CRAN (R 4.1.0)
##  sessioninfo            1.2.2    2021-12-06 [1] CRAN (R 4.1.0)
##  stringi                1.7.6    2021-11-29 [1] CRAN (R 4.1.0)
##  stringr              * 1.4.0    2019-02-10 [1] CRAN (R 4.1.0)
##  SummarizedExperiment * 1.24.0   2021-10-26 [1] Bioconductor
##  survival               3.2-13   2021-08-24 [1] CRAN (R 4.1.0)
##  testthat               3.1.1    2021-12-03 [1] CRAN (R 4.1.0)
##  tibble               * 3.1.6    2021-11-07 [1] CRAN (R 4.1.0)
##  tidyr                * 1.1.4    2021-09-27 [1] CRAN (R 4.1.0)
##  tidyselect             1.1.1    2021-04-30 [1] CRAN (R 4.1.0)
##  tidyverse            * 1.3.1    2021-04-15 [1] CRAN (R 4.1.0)
##  tzdb                   0.2.0    2021-10-27 [1] CRAN (R 4.1.0)
##  usethis                2.1.5    2021-12-09 [1] CRAN (R 4.1.0)
##  utf8                   1.2.2    2021-07-24 [1] CRAN (R 4.1.0)
##  vctrs                  0.3.8    2021-04-29 [1] CRAN (R 4.1.0)
##  viridisLite            0.4.0    2021-04-13 [1] CRAN (R 4.1.0)
##  withr                  2.4.3    2021-11-30 [1] CRAN (R 4.1.0)
##  xfun                   0.28     2021-11-04 [1] CRAN (R 4.1.0)
##  XML                    3.99-0.8 2021-09-17 [1] CRAN (R 4.1.0)
##  xml2                   1.3.3    2021-11-30 [1] CRAN (R 4.1.0)
##  xtable                 1.8-4    2019-04-21 [1] CRAN (R 4.1.0)
##  XVector                0.34.0   2021-10-26 [1] Bioconductor
##  yaml                   2.2.1    2020-02-01 [1] CRAN (R 4.1.0)
##  zlibbioc               1.40.0   2021-10-26 [1] Bioconductor
## 
##  [1] /Library/Frameworks/R.framework/Versions/4.1/Resources/library
## 
## ──────────────────────────────────────────────────────────────────────────────
</code></pre>

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../preprocessing/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Preprocessing" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Preprocessing
            </div>
          </div>
        </a>
      
      
        
        <a href="../DEA/" class="md-footer__link md-footer__link--next" aria-label="Next: Differential Expression" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Differential Expression
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../assets/javascripts/workers/search.bd0b6b67.min.js"}</script>
    
    
      <script src="../assets/javascripts/bundle.8aa65030.min.js"></script>
      
    
  </body>
</html>