
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.3, mkdocs-material-8.2.1">
    
    
      
        <title>Differential Expression - Bulk RNAseq data analysis - an introductory course</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.e8d9bf0c.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.e6a45f82.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../extra.css">
    
    <script>__md_scope=new URL("..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#differential-expression-analysis" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Bulk RNAseq data analysis - an introductory course" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis - an introductory course" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bulk RNAseq data analysis - an introductory course
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Differential Expression
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Bulk RNAseq data analysis - an introductory course" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis - an introductory course" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    Bulk RNAseq data analysis - an introductory course
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Course Introduction
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../preprocessing/" class="md-nav__link">
        Preprocessing
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../exploratory_analysis/" class="md-nav__link">
        Exploratory Analysis
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Differential Expression
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Differential Expression
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#extracting-results" class="md-nav__link">
    Extracting results
  </a>
  
    <nav class="md-nav" aria-label="Extracting results">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#save-significant-results" class="md-nav__link">
    Save significant results
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../contrast_design/" class="md-nav__link">
        Complex designs
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../functional_analysis/" class="md-nav__link">
        Functional Analysis
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#extracting-results" class="md-nav__link">
    Extracting results
  </a>
  
    <nav class="md-nav" aria-label="Extracting results">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#save-significant-results" class="md-nav__link">
    Save significant results
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                


<h1 id="differential-expression-analysis">Differential expression analysis</h1>
<p>In order to test for differential expression, we operate on raw counts
and use discrete distributions After checking that our samples are well
correlated within conditions, we can proceed to test for Differantially
Expressed Genes using the <em>results()</em> function.</p>
<pre><code>res &lt;- results(dds)
head(data.frame(res))

##                baseMean log2FoldChange     lfcSE        stat     pvalue
## FBgn0000008   95.144292    0.002276441 0.2237287  0.01017501 0.99188166
## FBgn0000014    1.056523   -0.495120386 2.1431858 -0.23102075 0.81729868
## FBgn0000017 4352.553569   -0.239918944 0.1263369 -1.89904085 0.05755911
## FBgn0000018  418.610484   -0.104673912 0.1484891 -0.70492676 0.48085582
## FBgn0000024    6.406200    0.210847792 0.6895876  0.30575928 0.75978794
## FBgn0000032  989.720217   -0.091788071 0.1477518 -0.62123130 0.53444745
##                  padj
## FBgn0000008 0.9972108
## FBgn0000014        NA
## FBgn0000017 0.2880017
## FBgn0000018 0.8268337
## FBgn0000024 0.9435011
## FBgn0000032 0.8542018
</code></pre>
<p>Information about the results columns can be retrieved using the
following snippet:</p>
<pre><code>mcols(res)$description

## [1] "mean of normalized counts for all samples"             
## [2] "log2 fold change (MLE): condition treated vs untreated"
## [3] "standard error: condition treated vs untreated"        
## [4] "Wald statistic: condition treated vs untreated"        
## [5] "Wald test p-value: condition treated vs untreated"     
## [6] "BH adjusted p-values"
</code></pre>
<p>In this example, we only have two conditions: “untreated” and “treated”,
so there is no need to specify which comparison we would like to make.
If your experiment contains more than two conditions, there are
different ways of specifying which comparison you would like to make.</p>
<p>First, the “contrast” argument will take a vector that specifies the
metadata and the order of the comparison. In this case, you can swap the
order of the condition levels.</p>
<pre><code>res &lt;- results(dds, contrast=c("condition","untreated","treated"))
head(data.frame(res))

##                baseMean log2FoldChange     lfcSE        stat     pvalue
## FBgn0000008   95.144292   -0.002276441 0.2237287 -0.01017501 0.99188166
## FBgn0000014    1.056523    0.495120386 2.1431858  0.23102075 0.81729868
## FBgn0000017 4352.553569    0.239918944 0.1263369  1.89904085 0.05755911
## FBgn0000018  418.610484    0.104673912 0.1484891  0.70492676 0.48085582
## FBgn0000024    6.406200   -0.210847792 0.6895876 -0.30575928 0.75978794
## FBgn0000032  989.720217    0.091788071 0.1477518  0.62123130 0.53444745
##                  padj
## FBgn0000008 0.9972108
## FBgn0000014        NA
## FBgn0000017 0.2880017
## FBgn0000018 0.8268337
## FBgn0000024 0.9435011
## FBgn0000032 0.8542018
</code></pre>
<p>Second, “name” argument will use the results of the <em>resultsNames()</em>
function. This function creates possible combinations of conditions, but
always using the reference condition for the comparison.</p>
<pre><code>resultsNames(dds)

## [1] "Intercept"                      "condition_treated_vs_untreated"

res &lt;- results(dds, name="condition_treated_vs_untreated")
head(data.frame(res))

##                baseMean log2FoldChange     lfcSE        stat     pvalue
## FBgn0000008   95.144292    0.002276441 0.2237287  0.01017501 0.99188166
## FBgn0000014    1.056523   -0.495120386 2.1431858 -0.23102075 0.81729868
## FBgn0000017 4352.553569   -0.239918944 0.1263369 -1.89904085 0.05755911
## FBgn0000018  418.610484   -0.104673912 0.1484891 -0.70492676 0.48085582
## FBgn0000024    6.406200    0.210847792 0.6895876  0.30575928 0.75978794
## FBgn0000032  989.720217   -0.091788071 0.1477518 -0.62123130 0.53444745
##                  padj
## FBgn0000008 0.9972108
## FBgn0000014        NA
## FBgn0000017 0.2880017
## FBgn0000018 0.8268337
## FBgn0000024 0.9435011
## FBgn0000032 0.8542018
</code></pre>
<p>The notebook <em>contrast_design.Rmd</em> contains information about how to
understand and define contrasts using model matrices, which might be a
more intuitive way of comparing samples or conditions.</p>
<h2 id="extracting-results">Extracting results</h2>
<p>The output of the <em>res()</em> function is the one you want to use/save to
identify differentially expressed genes using log2 Fold Change or
adjusted p-value thresholds.</p>
<p>There is a chance that the adjusted p-value is NA. If you will work on
this results, it might be helpful to change all NA to an adjusted
p-value of 1.</p>
<pre><code>res$padj[is.na(res$padj)] &lt;- 1
</code></pre>
<p>We can have a quick look of the adjusted p-value results by using the
following custom function:</p>
<pre><code>results_summary &lt;- function(x, alpha = 0.05, LFC = 1) {
  ngenes &lt;- nrow(x)
  signif &lt;- sum(x$padj &lt; alpha, na.rm = T)
  up &lt;- sum(x$padj &lt; alpha &amp; x$log2FoldChange &gt; LFC, na.rm = T)
  down &lt;- sum(x$padj &lt; alpha &amp; x$log2FoldChange &lt; -LFC, na.rm = T)
  results &lt;- c(paste0("Number of genes: ", ngenes),
    paste0("Number of genes with adjusted p-value &lt; ",alpha,": ", signif, " (", round((signif/ngenes)*100,digits = 2),"%)"),
    "  Of those:",
    paste0("    with LFC &lt; ", -LFC, ": ", down, " (", round((down/ngenes)*100,digits = 2),"%)"),
    paste0("    with LFC &gt; ", LFC, ": ", up, " (", round((up/ngenes)*100,digits = 2),"%)"))
  writeLines(paste(results, collapse = "\n"))
  return(paste(results, collapse = "\n"))
}

res_summary &lt;- results_summary(res, alpha = 0.05, LFC = 1)

## Number of genes: 9921
## Number of genes with adjusted p-value &lt; 0.05: 841 (8.48%)
##   Of those:
##     with LFC &lt; -1: 107 (1.08%)
##     with LFC &gt; 1: 115 (1.16%)
</code></pre>
<h3 id="save-significant-results">Save significant results</h3>
<p>Extracting results based on adjusted p-value and LFC. this way we will
get a filteres table with genes with LFC &gt; 1 or LFC &lt; -1 with an
adjusted p-value of 0.05</p>
<pre><code>LFC &lt;- 1
adj_pvalue &lt;- 0.05

sig_res &lt;- res[res$padj &lt; adj_pvalue &amp; abs(res$log2FoldChange) &gt; LFC,]
#write.table(x = sig_res, "./significant_results.tsv", quote = F, col.names = T, row.names = T, sep = "\t")
</code></pre>
<h1 id="dea-visualizations">DEA visualizations</h1>
<h2 id="count-matrix-heatmaps">Count matrix heatmaps</h2>
<p>We can visualize different heatmaps depending on what we are interested
on. We can see the top most expressed genes, the top most variable genes
and the DE genes.</p>
<h3 id="top-most-variable-genes">Top most variable genes</h3>
<pre><code>ntop &lt;- 20
select &lt;- order(rowSds(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:ntop]
df &lt;- as.data.frame(colData(dds)[,c("condition","type")])
pheatmap(assay(vsd)[select,], cluster_rows=T, show_rownames=FALSE, scale = "row",
         cluster_cols=FALSE, annotation_col=df)
</code></pre>
<p><img src="../DEA.Rmd/most_var_genes_heatmap-1.png" style="display: block; margin: auto;" /></p>
<h3 id="top-expressed-genes">Top expressed genes</h3>
<pre><code>ntop &lt;- 20
select &lt;- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:ntop]
df &lt;- as.data.frame(colData(dds)[,c("condition","type")])
pheatmap(assay(vsd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)
</code></pre>
<p><img src="../DEA.Rmd/top_exp_genes_heatmap-1.png" style="display: block; margin: auto;" /></p>
<h3 id="differentially-expressed-genes">Differentially expressed genes</h3>
<pre><code>select &lt;- rownames(sig_res)
df &lt;- as.data.frame(colData(dds)[,c("condition","type")])
pheatmap(assay(vsd)[rownames(vsd) %in% select,], cluster_rows=TRUE, show_rownames=FALSE, scale = "row",
         cluster_cols=FALSE, annotation_col=df)
</code></pre>
<p><img src="../DEA.Rmd/DEA_heatmap-1.png" style="display: block; margin: auto;" /></p>
<h2 id="fold-change-plots">Fold change plots</h2>
<p>For fold change plots, it is useful to shrunk log2 fold changes, which
removes the noise associated with log2 fold changes from low count genes
without requiring arbitrary filtering thresholds.</p>
<pre><code>resLFC &lt;- lfcShrink(dds, coef="condition_treated_vs_untreated", type="apeglm")
resLFC$padj[is.na(resLFC$padj)] &lt;- 1
</code></pre>
<h3 id="ma-plot">MA plot</h3>
<pre><code>alpha &lt;- 0.05

ggplot(data.frame(resLFC)) + theme_bw() + 
  labs(title = "MA plot", subtitle = res_summary,
       color = paste0("Adjusted p-value &lt; ",alpha)) +
  geom_point(aes(x = log10(baseMean), y = log2FoldChange, 
                 color = factor((padj &lt; alpha), levels = c("TRUE","FALSE"))), size = 1)
</code></pre>
<p><img src="../DEA.Rmd/MA_plot-1.png" style="display: block; margin: auto;" /></p>
<h3 id="volcano-plot">Volcano plot</h3>
<pre><code>alpha = 0.05
LFC = 1
ggplot(data.frame(resLFC)) + theme_bw() + 
  labs(color ="Significant genes", title = "Volcano plot", subtitle = res_summary) + 
  geom_hline(yintercept = -log10(alpha), linetype = "dashed") + 
  geom_vline(xintercept = c(-LFC,LFC), linetype = "dashed") + 
  geom_point(aes(x = log2FoldChange, y = -log10(padj),
                 color = factor((padj &lt; alpha &amp; abs(log2FoldChange) &gt; LFC), levels = c("TRUE","FALSE"))), size = 1)
</code></pre>
<p><img src="../DEA.Rmd/volcano_plot-1.png" style="display: block; margin: auto;" /></p>
<h3 id="glimma-plots">Glimma plots</h3>
<p>We can visualize our results interactively using the <code>glimmaMA()</code> and
<code>glimmaVolcano()</code> functions.</p>
<pre><code>glimma_MA &lt;- glimmaMA(dds, groups = dds$condition)
glimma_Volcano &lt;- glimmaVolcano(dds, groups = dds$condition)
</code></pre>
<p>Save your interactive plots using the <code>htmlwidgets::saveWidget()</code>
function.</p>
<pre><code>htmlwidgets::saveWidget(glimmaMA(dds, groups = dds$condition), "ma-plot.html")
htmlwidgets::saveWidget(glimmaVolcano(dds, groups = dds$condition), "volcano-plot.html")
</code></pre>
<h1 id="session-info">Session info</h1>
<p>Finally, we create a <code>session_info()</code> table that will allow anyone to
check what versions of R and packages are we using for reproducibility
purposes.</p>
<pre><code>devtools::session_info()

## ─ Session info ───────────────────────────────────────────────────────────────
##  setting  value
##  version  R version 4.1.0 (2021-05-18)
##  os       macOS Big Sur 10.16
##  system   x86_64, darwin17.0
##  ui       X11
##  language (EN)
##  collate  en_US.UTF-8
##  ctype    en_US.UTF-8
##  tz       Europe/Copenhagen
##  date     2022-02-25
##  pandoc   2.14.0.1 @ /usr/local/bin/ (via rmarkdown)
## 
## ─ Packages ───────────────────────────────────────────────────────────────────
##  package              * version    date (UTC) lib source
##  annotate               1.72.0     2021-10-26 [1] Bioconductor
##  AnnotationDbi          1.56.2     2021-11-09 [1] Bioconductor
##  apeglm                 1.16.0     2021-10-26 [1] Bioconductor
##  assertthat             0.2.1      2019-03-21 [1] CRAN (R 4.1.0)
##  backports              1.4.1      2021-12-13 [1] CRAN (R 4.1.0)
##  bbmle                  1.0.24     2021-08-05 [1] CRAN (R 4.1.0)
##  bdsmatrix              1.3-4      2020-01-13 [1] CRAN (R 4.1.0)
##  Biobase              * 2.54.0     2021-10-26 [1] Bioconductor
##  BiocGenerics         * 0.40.0     2021-10-26 [1] Bioconductor
##  BiocParallel           1.28.3     2021-12-09 [1] Bioconductor
##  Biostrings             2.62.0     2021-10-26 [1] Bioconductor
##  bit                    4.0.4      2020-08-04 [1] CRAN (R 4.1.0)
##  bit64                  4.0.5      2020-08-30 [1] CRAN (R 4.1.0)
##  bitops                 1.0-7      2021-04-24 [1] CRAN (R 4.1.0)
##  blob                   1.2.2      2021-07-23 [1] CRAN (R 4.1.0)
##  broom                  0.7.10     2021-10-31 [1] CRAN (R 4.1.0)
##  cachem                 1.0.6      2021-08-19 [1] CRAN (R 4.1.0)
##  callr                  3.7.0      2021-04-20 [1] CRAN (R 4.1.0)
##  cellranger             1.1.0      2016-07-27 [1] CRAN (R 4.1.0)
##  cli                    3.1.0      2021-10-27 [1] CRAN (R 4.1.0)
##  coda                   0.19-4     2020-09-30 [1] CRAN (R 4.1.0)
##  colorspace             2.0-2      2021-06-24 [1] CRAN (R 4.1.0)
##  crayon                 1.4.2      2021-10-29 [1] CRAN (R 4.1.0)
##  data.table             1.14.2     2021-09-27 [1] CRAN (R 4.1.0)
##  DBI                    1.1.1      2021-01-15 [1] CRAN (R 4.1.0)
##  dbplyr                 2.1.1      2021-04-06 [1] CRAN (R 4.1.0)
##  DelayedArray           0.20.0     2021-10-26 [1] Bioconductor
##  desc                   1.4.0      2021-09-28 [1] CRAN (R 4.1.0)
##  DESeq2               * 1.34.0     2021-10-26 [1] Bioconductor
##  devtools               2.4.3      2021-11-30 [1] CRAN (R 4.1.0)
##  digest                 0.6.29     2021-12-01 [1] CRAN (R 4.1.0)
##  dplyr                * 1.0.7      2021-06-18 [1] CRAN (R 4.1.0)
##  edgeR                  3.36.0     2021-10-26 [1] Bioconductor
##  ellipsis               0.3.2      2021-04-29 [1] CRAN (R 4.1.0)
##  emdbook                1.3.12     2020-02-19 [1] CRAN (R 4.1.0)
##  evaluate               0.14       2019-05-28 [1] CRAN (R 4.1.0)
##  fansi                  0.5.0      2021-05-25 [1] CRAN (R 4.1.0)
##  farver                 2.1.0      2021-02-28 [1] CRAN (R 4.1.0)
##  fastmap                1.1.0      2021-01-25 [1] CRAN (R 4.1.0)
##  forcats              * 0.5.1      2021-01-27 [1] CRAN (R 4.1.0)
##  fs                     1.5.2      2021-12-08 [1] CRAN (R 4.1.0)
##  genefilter             1.76.0     2021-10-26 [1] Bioconductor
##  geneplotter            1.72.0     2021-10-26 [1] Bioconductor
##  generics               0.1.1      2021-10-25 [1] CRAN (R 4.1.0)
##  GenomeInfoDb         * 1.30.0     2021-10-26 [1] Bioconductor
##  GenomeInfoDbData       1.2.7      2021-11-16 [1] Bioconductor
##  GenomicRanges        * 1.46.1     2021-11-18 [1] Bioconductor
##  ggplot2              * 3.3.5      2021-06-25 [1] CRAN (R 4.1.0)
##  Glimma               * 2.4.0      2021-10-26 [1] Bioconductor
##  glue                   1.5.1      2021-11-30 [1] CRAN (R 4.1.0)
##  gprofiler2           * 0.2.1      2021-08-23 [1] CRAN (R 4.1.0)
##  gtable                 0.3.0      2019-03-25 [1] CRAN (R 4.1.0)
##  haven                  2.4.3      2021-08-04 [1] CRAN (R 4.1.0)
##  highr                  0.9        2021-04-16 [1] CRAN (R 4.1.0)
##  hms                    1.1.1      2021-09-26 [1] CRAN (R 4.1.0)
##  htmltools              0.5.2      2021-08-25 [1] CRAN (R 4.1.0)
##  htmlwidgets            1.5.4      2021-09-08 [1] CRAN (R 4.1.0)
##  httr                   1.4.2      2020-07-20 [1] CRAN (R 4.1.0)
##  IRanges              * 2.28.0     2021-10-26 [1] Bioconductor
##  jsonlite               1.7.2      2020-12-09 [1] CRAN (R 4.1.0)
##  KEGGREST               1.34.0     2021-10-26 [1] Bioconductor
##  knitr                  1.36       2021-09-29 [1] CRAN (R 4.1.0)
##  labeling               0.4.2      2020-10-20 [1] CRAN (R 4.1.0)
##  lattice                0.20-45    2021-09-22 [1] CRAN (R 4.1.0)
##  lazyeval               0.2.2      2019-03-15 [1] CRAN (R 4.1.0)
##  lifecycle              1.0.1      2021-09-24 [1] CRAN (R 4.1.0)
##  limma                  3.50.0     2021-10-26 [1] Bioconductor
##  locfit                 1.5-9.4    2020-03-25 [1] CRAN (R 4.1.0)
##  lubridate              1.8.0      2021-10-07 [1] CRAN (R 4.1.0)
##  magrittr               2.0.1      2020-11-17 [1] CRAN (R 4.1.0)
##  MASS                   7.3-54     2021-05-03 [1] CRAN (R 4.1.0)
##  Matrix                 1.4-0      2021-12-08 [1] CRAN (R 4.1.0)
##  MatrixGenerics       * 1.6.0      2021-10-26 [1] Bioconductor
##  matrixStats          * 0.61.0     2021-09-17 [1] CRAN (R 4.1.0)
##  memoise                2.0.1      2021-11-26 [1] CRAN (R 4.1.0)
##  modelr                 0.1.8      2020-05-19 [1] CRAN (R 4.1.0)
##  munsell                0.5.0      2018-06-12 [1] CRAN (R 4.1.0)
##  mvtnorm                1.1-3      2021-10-08 [1] CRAN (R 4.1.0)
##  numDeriv               2016.8-1.1 2019-06-06 [1] CRAN (R 4.1.0)
##  pheatmap             * 1.0.12     2019-01-04 [1] CRAN (R 4.1.0)
##  pillar                 1.6.4      2021-10-18 [1] CRAN (R 4.1.0)
##  pkgbuild               1.3.0      2021-12-09 [1] CRAN (R 4.1.0)
##  pkgconfig              2.0.3      2019-09-22 [1] CRAN (R 4.1.0)
##  pkgload                1.2.4      2021-11-30 [1] CRAN (R 4.1.0)
##  plotly                 4.10.0     2021-10-09 [1] CRAN (R 4.1.0)
##  plyr                   1.8.6      2020-03-03 [1] CRAN (R 4.1.0)
##  png                    0.1-7      2013-12-03 [1] CRAN (R 4.1.0)
##  prettyunits            1.1.1      2020-01-24 [1] CRAN (R 4.1.0)
##  processx               3.5.2      2021-04-30 [1] CRAN (R 4.1.0)
##  ps                     1.6.0      2021-02-28 [1] CRAN (R 4.1.0)
##  purrr                * 0.3.4      2020-04-17 [1] CRAN (R 4.1.0)
##  R6                     2.5.1      2021-08-19 [1] CRAN (R 4.1.0)
##  RColorBrewer         * 1.1-2      2014-12-07 [1] CRAN (R 4.1.0)
##  Rcpp                   1.0.7      2021-07-07 [1] CRAN (R 4.1.0)
##  RCurl                  1.98-1.5   2021-09-17 [1] CRAN (R 4.1.0)
##  readr                * 2.1.1      2021-11-30 [1] CRAN (R 4.1.0)
##  readxl                 1.3.1      2019-03-13 [1] CRAN (R 4.1.0)
##  remotes                2.4.2      2021-11-30 [1] CRAN (R 4.1.0)
##  reprex                 2.0.1      2021-08-05 [1] CRAN (R 4.1.0)
##  rlang                  0.4.12     2021-10-18 [1] CRAN (R 4.1.0)
##  rmarkdown              2.11       2021-09-14 [1] CRAN (R 4.1.0)
##  rprojroot              2.0.2      2020-11-15 [1] CRAN (R 4.1.0)
##  RSQLite                2.2.9      2021-12-06 [1] CRAN (R 4.1.0)
##  rstudioapi             0.13       2020-11-12 [1] CRAN (R 4.1.0)
##  rvest                  1.0.2      2021-10-16 [1] CRAN (R 4.1.0)
##  S4Vectors            * 0.32.3     2021-11-21 [1] Bioconductor
##  scales                 1.1.1      2020-05-11 [1] CRAN (R 4.1.0)
##  sessioninfo            1.2.2      2021-12-06 [1] CRAN (R 4.1.0)
##  stringi                1.7.6      2021-11-29 [1] CRAN (R 4.1.0)
##  stringr              * 1.4.0      2019-02-10 [1] CRAN (R 4.1.0)
##  SummarizedExperiment * 1.24.0     2021-10-26 [1] Bioconductor
##  survival               3.2-13     2021-08-24 [1] CRAN (R 4.1.0)
##  testthat               3.1.1      2021-12-03 [1] CRAN (R 4.1.0)
##  tibble               * 3.1.6      2021-11-07 [1] CRAN (R 4.1.0)
##  tidyr                * 1.1.4      2021-09-27 [1] CRAN (R 4.1.0)
##  tidyselect             1.1.1      2021-04-30 [1] CRAN (R 4.1.0)
##  tidyverse            * 1.3.1      2021-04-15 [1] CRAN (R 4.1.0)
##  tzdb                   0.2.0      2021-10-27 [1] CRAN (R 4.1.0)
##  usethis                2.1.5      2021-12-09 [1] CRAN (R 4.1.0)
##  utf8                   1.2.2      2021-07-24 [1] CRAN (R 4.1.0)
##  vctrs                  0.3.8      2021-04-29 [1] CRAN (R 4.1.0)
##  viridisLite            0.4.0      2021-04-13 [1] CRAN (R 4.1.0)
##  withr                  2.4.3      2021-11-30 [1] CRAN (R 4.1.0)
##  xfun                   0.28       2021-11-04 [1] CRAN (R 4.1.0)
##  XML                    3.99-0.8   2021-09-17 [1] CRAN (R 4.1.0)
##  xml2                   1.3.3      2021-11-30 [1] CRAN (R 4.1.0)
##  xtable                 1.8-4      2019-04-21 [1] CRAN (R 4.1.0)
##  XVector                0.34.0     2021-10-26 [1] Bioconductor
##  yaml                   2.2.1      2020-02-01 [1] CRAN (R 4.1.0)
##  zlibbioc               1.40.0     2021-10-26 [1] Bioconductor
## 
##  [1] /Library/Frameworks/R.framework/Versions/4.1/Resources/library
## 
## ──────────────────────────────────────────────────────────────────────────────
</code></pre>

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../exploratory_analysis/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Exploratory Analysis" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Exploratory Analysis
            </div>
          </div>
        </a>
      
      
        
        <a href="../contrast_design/" class="md-footer__link md-footer__link--next" aria-label="Next: Complex designs" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Complex designs
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../assets/javascripts/workers/search.bd0b6b67.min.js"}</script>
    
    
      <script src="../assets/javascripts/bundle.8aa65030.min.js"></script>
      
    
  </body>
</html>