
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="img/sandbox1.png">
      <meta name="generator" content="mkdocs-1.3.0, mkdocs-material-8.2.14">
    
    
      
        <title>DESeq2 - Bulk RNAseq data analysis</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.3de6f41f.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.cc9b2e1e.min.css">
        
          
          
          <meta name="theme-color" content="#757575">
        
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="extra.css">
    
    <script>__md_scope=new URL(".",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="grey" data-md-color-accent="cyan">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-objectives" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="https://hds-sandbox.github.io/" title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component="logo">
      
  <img src="img/sandbox1.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bulk RNAseq data analysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              DESeq2
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="https://hds-sandbox.github.io/" title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component="logo">
      
  <img src="img/sandbox1.png" alt="logo">

    </a>
    Bulk RNAseq data analysis
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="index.html" class="md-nav__link">
        Course Introduction
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="02_setup.html" class="md-nav__link">
        Setup
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="03_experimental_planning.html" class="md-nav__link">
        Experimental planning
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="04_data_explanation.html" class="md-nav__link">
        Data Explanation
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="05_preprocessing.html" class="md-nav__link">
        Preprocessing
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          RNAseq counts
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="RNAseq counts" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          RNAseq counts
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="06a_count_matrix.html" class="md-nav__link">
        Count matrix
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="06b_count_normalization.html" class="md-nav__link">
        Normalization
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="07_exploratory_analysis.html" class="md-nav__link">
        Exploratory Analysis
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_8" type="checkbox" id="__nav_8" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_8">
          Differential Expression
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Differential Expression" data-md-level="1">
        <label class="md-nav__title" for="__nav_8">
          <span class="md-nav__icon md-icon"></span>
          Differential Expression
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08a_contrast_design.html" class="md-nav__link">
        Design formula
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          DESeq2
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="08b_DEA.html" class="md-nav__link md-nav__link--active">
        DESeq2
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    Learning Objectives
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#deseq2-differential-gene-expression-analysis-workflow" class="md-nav__link">
    DESeq2 differential gene expression analysis workflow
  </a>
  
    <nav class="md-nav" aria-label="DESeq2 differential gene expression analysis workflow">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-estimate-size-factors" class="md-nav__link">
    Step 1: Estimate size factors
  </a>
  
    <nav class="md-nav" aria-label="Step 1: Estimate size factors">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#mov10-de-analysis-examining-the-size-factors" class="md-nav__link">
    MOV10 DE analysis: examining the size factors
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-estimate-gene-wise-dispersion" class="md-nav__link">
    Step 2: Estimate gene-wise dispersion
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-3-fit-curve-to-gene-wise-dispersion-estimates" class="md-nav__link">
    Step 3: Fit curve to gene-wise dispersion estimates
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-4-shrink-gene-wise-dispersion-estimates-toward-the-values-predicted-by-the-curve" class="md-nav__link">
    Step 4: Shrink gene-wise dispersion estimates toward the values predicted by the curve
  </a>
  
    <nav class="md-nav" aria-label="Step 4: Shrink gene-wise dispersion estimates toward the values predicted by the curve">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#mov10-de-analysis-exploring-the-dispersion-estimates-and-assessing-model-fit" class="md-nav__link">
    MOV10 DE analysis: exploring the dispersion estimates and assessing model fit
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08c_hypothesis_testing.html" class="md-nav__link">
        Hypothesis testing
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08d_DEA_visualization.html" class="md-nav__link">
        Visualization
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_9" type="checkbox" id="__nav_9" >
      
      
      
      
        <label class="md-nav__link" for="__nav_9">
          Functional Analysis
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Functional Analysis" data-md-level="1">
        <label class="md-nav__title" for="__nav_9">
          <span class="md-nav__icon md-icon"></span>
          Functional Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="09a_FA_genomic_annotation.html" class="md-nav__link">
        Gene annotation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="09b_FA_overrepresentation.html" class="md-nav__link">
        Over represetation analysis
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="09c_FA_GSEA.html" class="md-nav__link">
        Class scoring
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="10_summarized_workflow.html" class="md-nav__link">
        Summarized workflow
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    Learning Objectives
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#deseq2-differential-gene-expression-analysis-workflow" class="md-nav__link">
    DESeq2 differential gene expression analysis workflow
  </a>
  
    <nav class="md-nav" aria-label="DESeq2 differential gene expression analysis workflow">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-estimate-size-factors" class="md-nav__link">
    Step 1: Estimate size factors
  </a>
  
    <nav class="md-nav" aria-label="Step 1: Estimate size factors">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#mov10-de-analysis-examining-the-size-factors" class="md-nav__link">
    MOV10 DE analysis: examining the size factors
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-estimate-gene-wise-dispersion" class="md-nav__link">
    Step 2: Estimate gene-wise dispersion
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-3-fit-curve-to-gene-wise-dispersion-estimates" class="md-nav__link">
    Step 3: Fit curve to gene-wise dispersion estimates
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-4-shrink-gene-wise-dispersion-estimates-toward-the-values-predicted-by-the-curve" class="md-nav__link">
    Step 4: Shrink gene-wise dispersion estimates toward the values predicted by the curve
  </a>
  
    <nav class="md-nav" aria-label="Step 4: Shrink gene-wise dispersion estimates toward the values predicted by the curve">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#mov10-de-analysis-exploring-the-dispersion-estimates-and-assessing-model-fit" class="md-nav__link">
    MOV10 DE analysis: exploring the dispersion estimates and assessing model fit
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                


  <h1>DESeq2</h1>

<p>Approximate time: 60 minutes</p>
<h2 id="learning-objectives">Learning Objectives</h2>
<ul>
<li>Explain the different steps involved in running <code>DESeq()</code></li>
<li>Examine size factors and undertand the source of differences</li>
<li>Inspect gene-level dispersion estimates</li>
<li>Recognize the importance of dispersion during differential
    expression analysis</li>
</ul>
<h2 id="deseq2-differential-gene-expression-analysis-workflow">DESeq2 differential gene expression analysis workflow</h2>
<p>Previously, we created the DESeq2 object using the appropriate design
formula and running DESeq2 using the two lines of code:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># DO NOT RUN</span>

<span class="c1">## Create DESeq2Dataset object</span>
<span class="n">dds</span> <span class="o">&lt;-</span> <span class="nf">DESeqDataSetFromMatrix</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">colData</span> <span class="o">=</span> <span class="n">meta</span><span class="p">,</span> <span class="n">design</span> <span class="o">=</span> <span class="o">~</span> <span class="n">sampletype</span><span class="p">)</span>

<span class="c1">## Run analysis</span>
<span class="n">dds</span> <span class="o">&lt;-</span> <span class="nf">DESeq</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
</code></pre></div>
<p>We completed the entire workflow for the differential gene expression
analysis with DESeq2. The steps in the analysis are output below:</p>
<p><img src="./img/08b_DEA/deseq2_workflow_separate.png" style="display: block; margin: auto;" /></p>
<p>We will be taking a detailed look at each of these steps to better
understand how DESeq2 is performing the statistical analysis and what
metrics we should examine to explore the quality of our analysis.</p>
<h3 id="step-1-estimate-size-factors">Step 1: Estimate size factors</h3>
<p>The first step in the differential expression analysis is to estimate
the size factors, which is exactly what we already did to normalize the
raw counts.</p>
<p><img src="./img/08b_DEA/deseq2_workflow_separate_sf.png" style="display: block; margin: auto;" /></p>
<p>DESeq2 will automatically estimate the size factors when performing the
differential expression analysis. However, if you have already generated
the size factors using <code>estimateSizeFactors()</code>, as we did earlier, then
DESeq2 will use these values.</p>
<p>To normalize the count data, DESeq2 calculates size factors for each
sample using the <em>median of ratios method</em> discussed previously in the
<a href="06b_count_normalization.html">‘Count normalization’</a> lesson.</p>
<h4 id="mov10-de-analysis-examining-the-size-factors">MOV10 DE analysis: examining the size factors</h4>
<p>Let’s take a quick look at size factor values we have for each sample:</p>
<div class="highlight"><pre><span></span><code>## Check the size factors
sizeFactors(dds)

Irrel_kd_1 Irrel_kd_2 Irrel_kd_3 Mov10_kd_2 Mov10_kd_3 Mov10_oe_1 Mov10_oe_2 
 1.1149694  0.9606733  0.7492240  1.5633640  0.9359695  1.2262649  1.1405026 
Mov10_oe_3 
 0.6542030
</code></pre></div>
<p>These numbers should be identical to those we generated initially when
we had run the function <code>estimateSizeFactors(dds)</code>. Take a look at the
total number of reads for each sample:</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Total number of raw counts per sample</span>
<span class="nf">colSums</span><span class="p">(</span><span class="nf">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">))</span>
</code></pre></div>
<p><em>How do the numbers correlate with the size factor?</em></p>
<p>We see that the larger size factors correspond to the samples with
higher sequencing depth, which makes sense, because to generate our
normalized counts we need to divide the counts by the size factors. This
accounts for the differences in sequencing depth between samples.</p>
<p>Now take a look at the total depth after normalization using:</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Total number of normalized counts per sample</span>
<span class="nf">colSums</span><span class="p">(</span><span class="nf">counts</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="bp">T</span><span class="p">))</span>
</code></pre></div>
<p><em>How do the values across samples compare with the total counts taken
for each sample?</em></p>
<p>You might have expected the counts to be the exact same across the
samples after normalization. However, DESeq2 also accounts for RNA
composition during the normalization procedure. By using the median
ratio value for the size factor, DESeq2 should not be biased to a large
number of counts sucked up by a few DE genes; however, this may lead to
the size factors being quite different than what would be anticipated
just based on sequencing depth.</p>
<h3 id="step-2-estimate-gene-wise-dispersion">Step 2: Estimate gene-wise dispersion</h3>
<p>The next step in the differential expression analysis is the estimation
of gene-wise dispersions. Before we get into the details, we should have
a good idea about what dispersion is referring to in DESeq2.</p>
<p><img src="./img/08b_DEA/deseq2_workflow_separate_dis.png" style="display: block; margin: auto;" /></p>
<p>In RNA-seq count data, we know:</p>
<ol>
<li>To determine differentially expressed genes, we need to identify
    genes that have significantly different mean expression between
    groups <strong>given the variation within the groups</strong> (between
    replicates).</li>
<li>The variation within group (between replicates) needs to account for
    the fact that variance increases with the mean expression, as shown
    in the plot below (each black dot is a gene).</li>
</ol>
<p><img src="./img/08b_DEA/deseq2_mean_vs_variance2.png" style="display: block; margin: auto;" /></p>
<p><strong>To accurately identify DE genes, DESeq2 needs to account for the
relationship between the variance and mean.</strong> We don’t want all of our
DE genes to be genes with low counts because the variance is lower for
lowly expressed genes.</p>
<p>Instead of using variance as the measure of variation in the data
(<em>since variance correlates with gene expression level</em>), DESeq2 uses a
measure of variation called <strong>dispersion, which accounts for a gene’s
variance and mean expression level</strong>. Dispersion is calculated by
<code>Var = μ + α*μ^2</code>, where <code>α</code> = dispersion, <code>Var</code> = variance, and <code>μ</code> =
mean, giving the relationship:</p>
<table>
<thead>
<tr>
<th align="center"></th>
<th align="center">Effect on dispersion</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">Variance increases</td>
<td align="center">Dispersion increases</td>
</tr>
<tr>
<td align="center">Mean expression increases</td>
<td align="center">Dispersion decreases</td>
</tr>
</tbody>
</table>
<p>For genes with moderate to high count values, the square root of
dispersion will be equal to the coefficient of variation. So 0.01
dispersion means 10% variation around the mean expected across
biological replicates. The dispersion estimates for genes with the same
mean will differ only based on their variance. <strong>Therefore, the
dispersion estimates reflect the variance in gene expression for a given
mean value.</strong> In the plot below, each black dot is a gene, and the
dispersion is plotted against the mean expression (across within-group
replicates) for each gene.</p>
<p><img src="./img/08b_DEA/deseq_dispersion1.png" style="display: block; margin: auto;" /></p>
<p><strong>How does the dispersion relate to our model?</strong></p>
<p>To accurately model sequencing counts, we need to generate accurate
estimates of within-group variation (variation between replicates of the
same sample group) for each gene. With only a few (3-6) replicates per
group, the <strong>estimates of variation for each gene are often
unreliable</strong>.</p>
<p>To address this problem, DESeq2 <strong>shares information across genes</strong> to
generate more accurate estimates of variation based on the mean
expression level of the gene using a method called ‘shrinkage’. <strong>DESeq2
assumes that genes with similar expression levels should have similar
dispersion.</strong></p>
<p><strong>Estimating the dispersion for each gene separately:</strong></p>
<p>DESeq2 estimates the dispersion for each gene based on the gene’s
expression level (mean counts of within-group replicates) and variance.</p>
<h3 id="step-3-fit-curve-to-gene-wise-dispersion-estimates">Step 3: Fit curve to gene-wise dispersion estimates</h3>
<p>The next step in the workflow is to fit a curve to the gene-wise
dispersion estimates. The idea behind fitting a curve to the data is
that different genes will have different scales of biological
variability, but, across all genes, there will be a distribution of
reasonable estimates of dispersion.</p>
<p><img src="./img/08b_DEA/deseq2_workflow_separate_fit.png" style="display: block; margin: auto;" /></p>
<p>This curve is displayed as a red line in the figure below, which plots
the estimate for the <strong>expected dispersion value for genes of a given
expression strength</strong>. Each black dot is a gene with an associated mean
expression level and maximum likelihood estimation (MLE) of the
dispersion (Step 1).</p>
<p><img src="./img/08b_DEA/deseq_dispersion1.png" style="display: block; margin: auto;" /></p>
<h3 id="step-4-shrink-gene-wise-dispersion-estimates-toward-the-values-predicted-by-the-curve">Step 4: Shrink gene-wise dispersion estimates toward the values predicted by the curve</h3>
<p>The next step in the workflow is to shrink the gene-wise dispersion
estimates toward the expected dispersion values.</p>
<p><img src="./img/08b_DEA/deseq2_workflow_separate_shr.png" style="display: block; margin: auto;" /></p>
<p>The curve allows for more accurate identification of differentially
expressed genes when sample sizes are small, and the strength of the
shrinkage for each gene depends on :</p>
<ul>
<li>how close gene dispersions are from the curve</li>
<li>sample size (more samples = less shrinkage)</li>
</ul>
<p><strong>This shrinkage method is particularly important to reduce false
positives in the differential expression analysis.</strong> Genes with low
dispersion estimates are shrunken towards the curve, and the more
accurate, higher shrunken values are output for fitting of the model and
differential expression testing. These shrunken estimates represent the
within-group variation that is needed to determine whether the gene
expression across groups is significantly different.</p>
<p>Dispersion estimates that are slightly above the curve are also shrunk
toward the curve for better dispersion estimation; however, genes with
<strong>extremely high dispersion values are not</strong>. This is due to the
likelihood that the gene does not follow the modeling assumptions and
has higher variability than others for biological or technical reasons
[<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">1</a>].
Shrinking the values toward the curve could result in false positives,
so these values are not shrunken. These genes are shown surrounded by
blue circles below.</p>
<p><img src="./img/08b_DEA/deseq_dispersion2.png" style="display: block; margin: auto;" /></p>
<p><strong>This is a good plot to evaluate whether your data is a good fit for
the DESeq2 model.</strong> You expect your data to generally scatter around the
curve, with the dispersion decreasing with increasing mean expression
levels. If you see a cloud or different shapes, then you might want to
explore your data more to see if you have contamination or outlier
samples. Note how much shrinkage you get across the whole range of means
in the <code>plotDispEsts()</code> plot for any experiment with low degrees of
freedom.</p>
<p>Examples of <strong>worrisome dispersion plots</strong> are shown below:</p>
<p>The plot below shows a cloud of dispersion values, which do not
generally follow the curve. This would suggest a bad fit of the data to
the model.</p>
<p><img src="./img/08b_DEA/bad_dispersion1.png" style="display: block; margin: auto;" /></p>
<p>The next plot shows the dispersion values initially decreasing, then
increasing with larger expression values. The larger mean expression
values should not have larger dispersions based on our expectations - we
expect decreasing dispersions with increasing mean. This indicates that
there is less variation for more highly expressed genes than expected.
This also indicates that there could be an outlier sample or
contamination present in our analysis.</p>
<p><img src="./img/08b_DEA/bad_dispersion2.png" style="display: block; margin: auto;" /></p>
<h4 id="mov10-de-analysis-exploring-the-dispersion-estimates-and-assessing-model-fit">MOV10 DE analysis: exploring the dispersion estimates and assessing model fit</h4>
<p>Let’s take a look at the dispersion estimates for our MOV10 data:</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Plot dispersion estimates</span>
<span class="nf">plotDispEsts</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
</code></pre></div>
<p><img src="./img/08b_DEA/plotDispersion.png" style="display: block; margin: auto;" /></p>
<p><strong>Since we have a small sample size, for many genes we see quite a bit
of shrinkage. Do you think our data are a good fit for the model?</strong></p>
<p>We see a nice decrease in dispersion with increasing mean expression,
which is good. We also see the dispersion estimates generally surround
the curve, which is also expected. Overall, this plot looks good. We do
see strong shrinkage, which is likely due to the fact that we have only
two replicates for one of our sample groups. The more replicates we
have, the less shrinkage is applied to the dispersion estimates, and the
more DE genes are able to be identified. We would generally recommend
having at least 4 biological replicates per condition for better
estimation of variation.</p>
<hr />
<p><strong>Exercise</strong></p>
<p>Given the dispersion plot below, would you have any concerns regarding
the fit of your data to the model?</p>
<ul>
<li>If not, what aspects of the plot makes you feel confident about your
    data?</li>
<li>If so, what are your concerns? What would you do to address them?</li>
</ul>
<p><img src="./img/08b_DEA/exercise_dispersion.png" style="display: block; margin: auto;" /></p>
<hr />
<p><em>This lesson was originally developed by members of the teaching team
(Mary Piper, Meeta Mistry, Radhika Khetani) at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan
Bioinformatics Core (HBC)</a>.</em></p>
<p><em>Some materials and hands-on activities were adapted from <a href="http://www.bioconductor.org/help/workflows/rnaseqGene/#de">RNA-seq
workflow</a> on
the Bioconductor website</em></p>
<hr />

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="08a_contrast_design.html" class="md-footer__link md-footer__link--prev" aria-label="Previous: Design formula" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Design formula
            </div>
          </div>
        </a>
      
      
        
        <a href="08c_hypothesis_testing.html" class="md-footer__link md-footer__link--next" aria-label="Next: Hypothesis testing" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Hypothesis testing
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": ".", "features": [], "search": "assets/javascripts/workers/search.2a1c317c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="assets/javascripts/bundle.c2e1ee47.min.js"></script>
      
    
  </body>
</html>