
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.3, mkdocs-material-8.2.5">
    
    
      
        <title>Over represetation analysis - Bulk RNAseq data analysis</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.2d9f7617.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.e6a45f82.min.css">
        
          
          
          <meta name="theme-color" content="#757575">
        
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="extra.css">
    
    <script>__md_scope=new URL(".",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="grey" data-md-color-accent="cyan">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-objectives" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="https://hds-sandbox.github.io/" title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component="logo">
      
  <img src="img/favicon.ico" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bulk RNAseq data analysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Over represetation analysis
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="https://hds-sandbox.github.io/" title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component="logo">
      
  <img src="img/favicon.ico" alt="logo">

    </a>
    Bulk RNAseq data analysis
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="index.html" class="md-nav__link">
        Course Introduction
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="02_setup.html" class="md-nav__link">
        Setup
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="03_experimental_planning.html" class="md-nav__link">
        Experimental planning
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="04_data_explanation.html" class="md-nav__link">
        Data Explanation
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="05_preprocessing.html" class="md-nav__link">
        Preprocessing
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          RNAseq counts
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="RNAseq counts" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          RNAseq counts
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="06a_count_matrix.html" class="md-nav__link">
        Count matrix
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="06b_count_normalization.html" class="md-nav__link">
        Normalization
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="07_exploratory_analysis.html" class="md-nav__link">
        Exploratory Analysis
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_8" type="checkbox" id="__nav_8" >
      
      
      
      
        <label class="md-nav__link" for="__nav_8">
          Differential Expression
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Differential Expression" data-md-level="1">
        <label class="md-nav__title" for="__nav_8">
          <span class="md-nav__icon md-icon"></span>
          Differential Expression
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08a_contrast_design.html" class="md-nav__link">
        Design formula
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08b_DEA.html" class="md-nav__link">
        DESeq2
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08c_hypothesis_testing.html" class="md-nav__link">
        Hypothesis testing
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08d_DEA_visualization.html" class="md-nav__link">
        Visualization
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_9" type="checkbox" id="__nav_9" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_9">
          Functional Analysis
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Functional Analysis" data-md-level="1">
        <label class="md-nav__title" for="__nav_9">
          <span class="md-nav__icon md-icon"></span>
          Functional Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="09a_FA_genomic_annotation.html" class="md-nav__link">
        Gene annotation
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Over represetation analysis
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="09b_FA_overrepresentation.html" class="md-nav__link md-nav__link--active">
        Over represetation analysis
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    Learning Objectives:
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#functional-analysis" class="md-nav__link">
    Functional analysis
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#over-representation-analysis" class="md-nav__link">
    Over-representation analysis
  </a>
  
    <nav class="md-nav" aria-label="Over-representation analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#hypergeometric-testing" class="md-nav__link">
    Hypergeometric testing
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gene-ontology-project" class="md-nav__link">
    Gene Ontology project
  </a>
  
    <nav class="md-nav" aria-label="Gene Ontology project">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#go-ontologies" class="md-nav__link">
    GO Ontologies
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#go-term-hierarchy" class="md-nav__link">
    GO term hierarchy
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#clusterprofiler" class="md-nav__link">
    clusterProfiler
  </a>
  
    <nav class="md-nav" aria-label="clusterProfiler">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#running-clusterprofiler" class="md-nav__link">
    Running clusterProfiler
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualizing-clusterprofiler-results" class="md-nav__link">
    Visualizing clusterProfiler results
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="09c_FA_GSEA.html" class="md-nav__link">
        Class scoring
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="10_summarized_workflow.html" class="md-nav__link">
        Summarized workflow
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    Learning Objectives:
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#functional-analysis" class="md-nav__link">
    Functional analysis
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#over-representation-analysis" class="md-nav__link">
    Over-representation analysis
  </a>
  
    <nav class="md-nav" aria-label="Over-representation analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#hypergeometric-testing" class="md-nav__link">
    Hypergeometric testing
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gene-ontology-project" class="md-nav__link">
    Gene Ontology project
  </a>
  
    <nav class="md-nav" aria-label="Gene Ontology project">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#go-ontologies" class="md-nav__link">
    GO Ontologies
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#go-term-hierarchy" class="md-nav__link">
    GO term hierarchy
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#clusterprofiler" class="md-nav__link">
    clusterProfiler
  </a>
  
    <nav class="md-nav" aria-label="clusterProfiler">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#running-clusterprofiler" class="md-nav__link">
    Running clusterProfiler
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#visualizing-clusterprofiler-results" class="md-nav__link">
    Visualizing clusterProfiler results
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                


  <h1>Over represetation analysis</h1>

<p>Approximate time: 120 minutes</p>
<h2 id="learning-objectives">Learning Objectives:</h2>
<ul>
<li>Determine how functions are attributed to genes using Gene Ontology
    terms</li>
<li>Describe the theory of how functional enrichment tools yield
    statistically enriched functions or interactions</li>
<li>Discuss functional analysis using over-representation analysis,
    functional class scoring, and pathway topology methods</li>
<li>Identify popular functional analysis tools for over-representation
    analysis</li>
</ul>
<h2 id="functional-analysis">Functional analysis</h2>
<p>The output of RNA-seq differential expression analysis is a list of
significant differentially expressed genes (DEGs). To gain greater
biological insight on the differentially expressed genes there are
various analyses that can be done:</p>
<ul>
<li>determine whether there is enrichment of known biological functions,
    interactions, or pathways</li>
<li>identify genes’ involvement in novel pathways or networks by
    grouping genes together based on similar trends</li>
<li>use global changes in gene expression by visualizing all genes being
    significantly up- or down-regulated in the context of external
    interaction data</li>
</ul>
<p>Generally for any differential expression analysis, it is useful to
interpret the resulting gene lists using freely available web- and
R-based tools. While tools for functional analysis span a wide variety
of techniques, they can loosely be categorized into three main types:
over-representation analysis, functional class scoring, and pathway
topology
[<a href="https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course/raw/master/resources/pathway_tools.pdf">1</a>].</p>
<p><img src="./img/09b_FA_overrepresentation/pathway_analysis.png" style="display: block; margin: auto;" /></p>
<p>The goal of functional analysis is provide biological insight, so it’s
necessary to analyze our results in the context of our experimental
hypothesis: <strong>FMRP and MOV10 associate and regulate the translation of a
subset of RNAs</strong>. Therefore, based on the authors’ hypothesis, we may
expect the enrichment of processes/pathways related to <strong>translation,
splicing, and the regulation of mRNAs</strong>, which we would need to validate
experimentally.</p>
<p><strong><em>Note that all tools described below are great tools to validate
experimental results and to make hypotheses. These tools suggest
genes/pathways that may be involved with your condition of interest;
however, you should NOT use these tools to make conclusions about the
pathways involved in your experimental process. You will need to perform
experimental validation of any suggested pathways.</em></strong></p>
<h2 id="over-representation-analysis">Over-representation analysis</h2>
<p>There are a plethora of functional enrichment tools that perform some
type of “over-representation” analysis by querying databases containing
information about gene function and interactions.</p>
<p>These databases typically <strong>categorize genes into groups (gene sets)</strong>
based on shared function, or involvement in a pathway, or presence in a
specific cellular location, or other categorizations, e.g. functional
pathways, etc. Essentially, known genes are binned into categories that
have been consistently named (controlled vocabulary) based on how the
gene has been annotated functionally. These categories are independent
of any organism, however each organism has distinct categorizations
available.</p>
<p>To determine whether any categories are over-represented, you can
determine the probability of having the observed proportion of genes
associated with a specific category in your gene list based on the
proportion of genes associated with the same category in the background
set (gene categorizations for the appropriate organism).</p>
<p><img src="./img/09b_FA_overrepresentation/go_proportions.png" style="display: block; margin: auto;" /></p>
<p><img src="./img/09b_FA_overrepresentation/go_proportions_table3.png" style="display: block; margin: auto;" /></p>
<p>The statistical test that will determine whether something is actually
over-represented is the <em>Hypergeometric test</em>.</p>
<h3 id="hypergeometric-testing">Hypergeometric testing</h3>
<p>Using the example of the first functional category above, hypergeometric
distribution is a probability distribution that describes the
probability of 25 genes (k) being associated with “Functional category
1”, for all genes in our gene list (n=1000), from a population of all of
the genes in entire genome (N=13,000) which contains 35 genes (K)
associated with “Functional category 1”
[<a href="https://en.wikipedia.org/wiki/Hypergeometric_distribution">4</a>].</p>
<p>The calculation of probability of k successes follows the formula:</p>
<p><img src="./img/09b_FA_overrepresentation/hypergeo.png" style="display: block; margin: auto;" /></p>
<p>This test will result in an adjusted p-value (after multiple test
correction) for each category tested.</p>
<h2 id="gene-ontology-project">Gene Ontology project</h2>
<p>One of the most widely-used categorizations is the <strong>Gene Ontology
(GO)</strong> established by the Gene Ontology project.</p>
<p>“The Gene Ontology project is a collaborative effort to address the need
for consistent descriptions of gene products across databases”
[<a href="http://geneontology.org/page/introduction-go-resource">2</a>]. The
<a href="http://geneontology.org/page/go-consortium-contributors-list">Gene Ontology
Consortium</a>
maintains the GO terms, and these GO terms are incorporated into gene
annotations in many of the popular repositories for animal, plant, and
microbial genomes.</p>
<p>Tools that investigate enrichment of biological functions or
interactions often use the Gene Ontology (GO) categorizations, i.e. the
GO terms to determine whether any have significantly modified
representation in a given list of genes. Therefore, to best use and
interpret the results from these functional analysis tools, it is
helpful to have a good understanding of the GO terms themselves and
their organization.</p>
<h3 id="go-ontologies">GO Ontologies</h3>
<p>To describe the roles of genes and gene products, GO terms are organized
into three independent controlled vocabularies (ontologies) in a
species-independent manner:</p>
<ul>
<li><strong>Biological process:</strong> refers to the biological role involving the
    gene or gene product, and could include “transcription”, “signal
    transduction”, and “apoptosis”. A biological process generally
    involves a chemical or physical change of the starting material or
    input.</li>
<li><strong>Molecular function:</strong> represents the biochemical activity of the
    gene product, such activities could include “ligand”, “GTPase”, and
    “transporter”.</li>
<li><strong>Cellular component:</strong> refers to the location in the cell of the
    gene product. Cellular components could include “nucleus”,
    “lysosome”, and “plasma membrane”.</li>
</ul>
<p>Each GO term has a term name (e.g. <strong>DNA repair</strong>) and a unique term
accession number (<strong><GO:0005125></strong>), and a single gene product can be
associated with many GO terms, since a single gene product “may function
in several processes, contain domains that carry out diverse molecular
functions, and participate in multiple alternative interactions with
other proteins, organelles or locations in the cell”
[<a href="https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course/raw/master/resources/go_terms.pdf">3</a>].</p>
<h3 id="go-term-hierarchy">GO term hierarchy</h3>
<p>Some gene products are well-researched, with vast quantities of data
available regarding their biological processes and functions. However,
other gene products have very little data available about their roles in
the cell.</p>
<p>For example, the protein, “p53”, would contain a wealth of information
on it’s roles in the cell, whereas another protein might only be known
as a “membrane-bound protein” with no other information available.</p>
<p>The GO ontologies were developed to describe and query biological
knowledge with differing levels of information available. To do this, GO
ontologies are loosely hierarchical, ranging from general, ‘parent’,
terms to more specific, ‘child’ terms. The GO ontologies are “loosely”
hierarchical since ‘child’ terms can have multiple ‘parent’ terms.</p>
<p>Some genes with less information may only be associated with general
‘parent’ terms or no terms at all, while other genes with a lot of
information be associated with many terms.</p>
<p><img alt="Nature Reviews Cancer 7, 23-34 (January
2007)" src="img/09b_FA_overrepresentation/go_heirarchy.jpg" /></p>
<p><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003343">Tips for working with GO
terms</a></p>
<h2 id="clusterprofiler">clusterProfiler</h2>
<p>We will be using
<a href="http://bioconductor.org/packages/release/bioc/html/clusterProfiler.html">clusterProfiler</a>
to perform over-representation analysis on GO terms associated with our
list of significant genes. The tool takes as input a significant gene
list and a background gene list and performs statistical enrichment
analysis using hypergeometric testing. The basic arguments allow the
user to select the appropriate organism and GO ontology (BP, CC, MF) to
test.</p>
<h3 id="running-clusterprofiler">Running clusterProfiler</h3>
<p>To run clusterProfiler GO over-representation analysis, we will change
our gene names into Ensembl IDs, since the tool works a bit easier with
the Ensembl IDs.</p>
<p>Then load the following libraries:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Load libraries</span>
<span class="nf">library</span><span class="p">(</span><span class="n">DOSE</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">pathview</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">clusterProfiler</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">)</span>
</code></pre></div>
<p>For the different steps in the functional analysis, we require Ensembl
and Entrez IDs. We will use the gene annotations that we generated
previously to merge with our differential expression results.</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Merge the AnnotationHub dataframe with the results </span>
<span class="n">res_ids</span> <span class="o">&lt;-</span> <span class="nf">left_join</span><span class="p">(</span><span class="n">res_tableOE_tb</span><span class="p">,</span> <span class="n">annotations_ahb</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;gene&quot;</span><span class="o">=</span><span class="s">&quot;gene_id&quot;</span><span class="p">))</span>    
</code></pre></div>
<blockquote>
<p><em><strong>NOTE:</strong> If you were unable to generate the <code>annotations_ahb</code>
object, you can download the annotations to your <code>data</code> folder by
right-clicking
<a href="https://github.com/hbctraining/DGE_workshop_salmon_online/raw/master/data/annotations_ahb.csv">here</a>
and selecting “Save link as…”</em></p>
<p><em>To read in the object, you can run the following code:</em>
<code>annotations_ahb &lt;- read.csv("annotations_ahb.csv")</code></p>
</blockquote>
<p>To perform the over-representation analysis, we need a list of
background genes and a list of significant genes. For our background
dataset we will use all genes tested for differential expression (all
genes in our results table). For our significant gene list we will use
genes with p-adjusted values less than 0.05 (we could include a fold
change threshold too if we have many DE genes).</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Create background dataset for hypergeometric testing using all genes tested for significance in the results                 </span>
<span class="n">allOE_genes</span> <span class="o">&lt;-</span> <span class="nf">as.character</span><span class="p">(</span><span class="n">res_ids</span><span class="o">$</span><span class="n">gene</span><span class="p">)</span>

<span class="c1">## Extract significant results</span>
<span class="n">sigOE</span> <span class="o">&lt;-</span> <span class="n">dplyr</span><span class="o">::</span><span class="nf">filter</span><span class="p">(</span><span class="n">res_ids</span><span class="p">,</span> <span class="n">padj</span> <span class="o">&lt;</span> <span class="m">0.05</span><span class="p">)</span>

<span class="n">sigOE_genes</span> <span class="o">&lt;-</span> <span class="nf">as.character</span><span class="p">(</span><span class="n">sigOE</span><span class="o">$</span><span class="n">gene</span><span class="p">)</span>
</code></pre></div>
<p>Now we can perform the GO enrichment analysis and save the results:</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Run GO enrichment analysis </span>
<span class="n">ego</span> <span class="o">&lt;-</span> <span class="nf">enrichGO</span><span class="p">(</span><span class="n">gene</span> <span class="o">=</span> <span class="n">sigOE_genes</span><span class="p">,</span> 
                <span class="n">universe</span> <span class="o">=</span> <span class="n">allOE_genes</span><span class="p">,</span>
                <span class="n">keyType</span> <span class="o">=</span> <span class="s">&quot;ENSEMBL&quot;</span><span class="p">,</span>
                <span class="n">OrgDb</span> <span class="o">=</span> <span class="n">org.Hs.eg.db</span><span class="p">,</span> 
                <span class="n">ont</span> <span class="o">=</span> <span class="s">&quot;BP&quot;</span><span class="p">,</span> 
                <span class="n">pAdjustMethod</span> <span class="o">=</span> <span class="s">&quot;BH&quot;</span><span class="p">,</span> 
                <span class="n">qvalueCutoff</span> <span class="o">=</span> <span class="m">0.05</span><span class="p">,</span> 
                <span class="n">readable</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
<blockquote>
<p><strong>NOTE:</strong> The different organisms with annotation databases available
to use with for the <code>OrgDb</code> argument can be found
<a href="img/09b_FA_overrepresentation/orgdb_annotation_databases.png">here</a>.</p>
<p>Also, the <code>keyType</code> argument may be coded as <code>keytype</code> in different
versions of clusterProfiler.</p>
<p>Finally, the <code>ont</code> argument can accept either “BP” (Biological
Process), “MF” (Molecular Function), and “CC” (Cellular Component)
subontologies, or “ALL” for all three.</p>
</blockquote>
<div class="highlight"><pre><span></span><code><span class="c1">## Output results from GO analysis to a table</span>
<span class="n">cluster_summary</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">ego</span><span class="p">)</span>

<span class="nf">write.csv</span><span class="p">(</span><span class="n">cluster_summary</span><span class="p">,</span> <span class="s">&quot;results/clusterProfiler_Mov10oe.csv&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img src="./img/09b_FA_overrepresentation/cluster_summary.png" style="display: block; margin: auto;" /></p>
<blockquote>
<p><strong>NOTE:</strong> Instead of saving just the results summary from the <code>ego</code>
object, it might also be beneficial to save the object itself. The
<code>save()</code> function enables you to save it as a <code>.rda</code> file,
e.g. <code>save(ego, file="results/ego.rda")</code>. The complementary function
to <code>save()</code> is the function <code>load()</code>,
e.g. <code>ego &lt;- load(file="results/ego.rda")</code>.</p>
<p><em>This is a useful set of functions to know, since it enables one to
preserve analyses at specific stages and reload them when needed.</em>
More information about these functions can be found
<a href="https://www.r-bloggers.com/load-save-and-rda-files/">here</a> &amp;
<a href="http://rpubs.com/euclid/387778">here</a>.</p>
</blockquote>
<hr />
<blockquote>
<p><strong>NOTE:</strong> <strong>You can also perform GO enrichment analysis with only the
up or down regulated genes</strong> in addition to performing it for the full
list of significant genes. This can be useful to identify GO terms
impacted in one direction and not the other. If very few genes are in
any of these lists (\&lt; 50, roughly) it may not be possible to get any
significant GO terms.</p>
<div class="highlight"><pre><span></span><code>## Extract upregulated genes
sigOE_up &lt;- dplyr::filter(res_ids, padj &lt; 0.05 &amp; log2foldchange &gt; 0)

sigOE_up_genes &lt;- as.character(sigOE_up$gene)

## Extract downregulated genes
sigOE_down &lt;- dplyr::filter(res_ids, padj &lt; 0.05 &amp; log2foldchange &lt; 0)

sigOE_down_genes &lt;- as.character(sigOE_down$gene)
</code></pre></div>
<p>You can then create <code>ego_up</code> &amp; <code>ego_down</code> objects by running the
<code>enrichGO()</code> function for <code>gene = sigOE_up_genes</code> or
<code>gene = sigOE_down_genes</code>.</p>
</blockquote>
<hr />
<h3 id="visualizing-clusterprofiler-results">Visualizing clusterProfiler results</h3>
<p>clusterProfiler has a variety of options for viewing the
over-represented GO terms. We will explore the dotplot, enrichment plot,
and the category netplot.</p>
<p>The <strong>dotplot</strong> shows the number of genes associated with the first 50
terms (size) and the p-adjusted values for these terms (color). This
plot displays the top 50 GO terms by gene ratio (# genes related to GO
term / total number of sig genes), not p-adjusted value.</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Dotplot </span>
<span class="nf">dotplot</span><span class="p">(</span><span class="n">ego</span><span class="p">,</span> <span class="n">showCategory</span><span class="o">=</span><span class="m">50</span><span class="p">)</span>
</code></pre></div>
<p><strong>To save the figure,</strong> click on the <code>Export</code> button in the RStudio
<code>Plots</code> tab and <code>Save as PDF...</code>. In the pop-up window, change: -
<code>Orientation:</code> to <code>Landscape</code> - <code>PDF size</code> to <code>8 x 14</code> to give a figure
of appropriate size for the text labels</p>
<p><img src="./img/09b_FA_overrepresentation/mov10oe_dotplot.png" style="display: block; margin: auto;" /></p>
<p>The next plot is the <strong>enrichment GO plot</strong>, which shows the
relationship between the top 50 most significantly enriched GO terms
(padj.), by grouping similar terms together. Before creating the plot,
we will need to obtain the similarity between terms using the
<code>pairwise_termsim()</code> function (<a href="https://rdrr.io/github/GuangchuangYu/enrichplot/man/emapplot.html">instructions for
emapplot</a>).
In the enrichment plot, the color represents the p-values relative to
the other displayed terms (brighter red is more significant), and the
size of the terms represents the number of genes that are significant
from our list.</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Add similarity matrix to the termsim slot of enrichment result</span>
<span class="n">ego</span> <span class="o">&lt;-</span> <span class="n">enrichplot</span><span class="o">::</span><span class="nf">pairwise_termsim</span><span class="p">(</span><span class="n">ego</span><span class="p">)</span>

<span class="c1">## Enrichmap clusters the 50 most significant (by padj) GO terms to visualize relationships between terms</span>
<span class="nf">emapplot</span><span class="p">(</span><span class="n">ego</span><span class="p">,</span> <span class="n">showCategory</span> <span class="o">=</span> <span class="m">50</span><span class="p">)</span>
</code></pre></div>
<p><strong>To save the figure,</strong> click on the <code>Export</code> button in the RStudio
<code>Plots</code> tab and <code>Save as PDF...</code>. In the pop-up window, change the
<code>PDF size</code> to <code>12 x 14</code> to give a figure of appropriate size for the
text labels.</p>
<p><img src="./img/09b_FA_overrepresentation/emapplot.png" style="display: block; margin: auto;" /></p>
<p>Finally, the <strong>category netplot</strong> shows the relationships between the
genes associated with the top five most significant GO terms and the
fold changes of the significant genes associated with these terms
(color). The size of the GO terms reflects the pvalues of the terms,
with the more significant terms being larger. This plot is particularly
useful for hypothesis generation in identifying genes that may be
important to several of the most affected processes.</p>
<blockquote>
<p><strong>Note</strong> - You may need to install the <code>ggnewscale</code> package using
<code>install.packages("ggnewscale")</code> for the <code>cnetplot()</code> function to
work.</p>
</blockquote>
<div class="highlight"><pre><span></span><code><span class="c1">## To color genes by log2 fold changes, we need to extract the log2 fold changes from our results table creating a named vector</span>
<span class="n">OE_foldchanges</span> <span class="o">&lt;-</span> <span class="n">sigOE</span><span class="o">$</span><span class="n">log2FoldChange</span>

<span class="nf">names</span><span class="p">(</span><span class="n">OE_foldchanges</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">sigOE</span><span class="o">$</span><span class="n">gene</span>

<span class="c1">## Cnetplot details the genes associated with one or more terms - by default gives the top 5 significant terms (by padj)</span>
<span class="nf">cnetplot</span><span class="p">(</span><span class="n">ego</span><span class="p">,</span> 
         <span class="n">categorySize</span><span class="o">=</span><span class="s">&quot;pvalue&quot;</span><span class="p">,</span> 
         <span class="n">showCategory</span> <span class="o">=</span> <span class="m">5</span><span class="p">,</span> 
         <span class="n">foldChange</span><span class="o">=</span><span class="n">OE_foldchanges</span><span class="p">,</span> 
         <span class="n">vertex.label.font</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>

<span class="c1">## If some of the high fold changes are getting drowned out due to a large range, you could set a maximum fold change value</span>
<span class="n">OE_foldchanges</span> <span class="o">&lt;-</span> <span class="nf">ifelse</span><span class="p">(</span><span class="n">OE_foldchanges</span> <span class="o">&gt;</span> <span class="m">2</span><span class="p">,</span> <span class="m">2</span><span class="p">,</span> <span class="n">OE_foldchanges</span><span class="p">)</span>
<span class="n">OE_foldchanges</span> <span class="o">&lt;-</span> <span class="nf">ifelse</span><span class="p">(</span><span class="n">OE_foldchanges</span> <span class="o">&lt;</span> <span class="m">-2</span><span class="p">,</span> <span class="m">-2</span><span class="p">,</span> <span class="n">OE_foldchanges</span><span class="p">)</span>

<span class="nf">cnetplot</span><span class="p">(</span><span class="n">ego</span><span class="p">,</span> 
         <span class="n">categorySize</span><span class="o">=</span><span class="s">&quot;pvalue&quot;</span><span class="p">,</span> 
         <span class="n">showCategory</span> <span class="o">=</span> <span class="m">5</span><span class="p">,</span> 
         <span class="n">foldChange</span><span class="o">=</span><span class="n">OE_foldchanges</span><span class="p">,</span> 
         <span class="n">vertex.label.font</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
</code></pre></div>
<p><strong>Again, to save the figure,</strong> click on the <code>Export</code> button in the
RStudio <code>Plots</code> tab and <code>Save as PDF...</code>. Change the <code>PDF size</code> to
<code>12 x 14</code> to give a figure of appropriate size for the text labels.</p>
<p><img src="./img/09b_FA_overrepresentation/cnetplot1.png" style="display: block; margin: auto;" /></p>
<p>If you are interested in significant processes that are <strong>not</strong> among
the top five, you can subset your <code>ego</code> dataset to only display these
processes:</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Subsetting the ego results without overwriting original `ego` variable</span>
<span class="n">ego2</span> <span class="o">&lt;-</span> <span class="n">ego</span>

<span class="n">ego2</span><span class="o">@</span><span class="n">result</span> <span class="o">&lt;-</span> <span class="n">ego</span><span class="o">@</span><span class="n">result</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">3</span><span class="p">,</span><span class="m">4</span><span class="p">,</span><span class="m">8</span><span class="p">,</span><span class="m">9</span><span class="p">),]</span>

<span class="c1">## Plotting terms of interest</span>
<span class="nf">cnetplot</span><span class="p">(</span><span class="n">ego2</span><span class="p">,</span> 
         <span class="n">categorySize</span><span class="o">=</span><span class="s">&quot;pvalue&quot;</span><span class="p">,</span> 
         <span class="n">foldChange</span><span class="o">=</span><span class="n">OE_foldchanges</span><span class="p">,</span> 
         <span class="n">showCategory</span> <span class="o">=</span> <span class="m">5</span><span class="p">,</span> 
         <span class="n">vertex.label.font</span><span class="o">=</span><span class="m">6</span><span class="p">)</span>
</code></pre></div>
<p><img src="./img/09b_FA_overrepresentation/cnetplot-2.png" style="display: block; margin: auto;" /></p>
<hr />
<p><em>This lesson was originally developed by members of the teaching team
(Mary Piper, Radhika Khetani) at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core
(HBC)</a>.</em></p>

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="09a_FA_genomic_annotation.html" class="md-footer__link md-footer__link--prev" aria-label="Previous: Gene annotation" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Gene annotation
            </div>
          </div>
        </a>
      
      
        
        <a href="09c_FA_GSEA.html" class="md-footer__link md-footer__link--next" aria-label="Next: Class scoring" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Class scoring
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": ".", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "assets/javascripts/workers/search.bd0b6b67.min.js"}</script>
    
    
      <script src="assets/javascripts/bundle.467223ff.min.js"></script>
      
    
  </body>
</html>