
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.3, mkdocs-material-8.2.5">
    
    
      
        <title>Class scoring - Bulk RNAseq data analysis - an introductory course</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.2d9f7617.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.e6a45f82.min.css">
        
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="extra.css">
    
    <script>__md_scope=new URL(".",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#functional-class-scoring" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="index.html" title="Bulk RNAseq data analysis - an introductory course" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis - an introductory course" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bulk RNAseq data analysis - an introductory course
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Class scoring
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="index.html" title="Bulk RNAseq data analysis - an introductory course" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis - an introductory course" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    Bulk RNAseq data analysis - an introductory course
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="index.html" class="md-nav__link">
        Course Introduction
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="02_setup.html" class="md-nav__link">
        Setup
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="03_experimental_planning.html" class="md-nav__link">
        Experimental planning
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="04_data_explanation.html" class="md-nav__link">
        Data Explanation
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="05_preprocessing.html" class="md-nav__link">
        Preprocessing
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          RNAseq counts
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="RNAseq counts" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          RNAseq counts
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="06a_count_matrix.html" class="md-nav__link">
        Count matrix
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="06b_count_normalization.html" class="md-nav__link">
        Normalization
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="07_exploratory_analysis.html" class="md-nav__link">
        Exploratory Analysis
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_8" type="checkbox" id="__nav_8" >
      
      
      
      
        <label class="md-nav__link" for="__nav_8">
          Differential Expression
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Differential Expression" data-md-level="1">
        <label class="md-nav__title" for="__nav_8">
          <span class="md-nav__icon md-icon"></span>
          Differential Expression
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08a_contrast_design.html" class="md-nav__link">
        Design formula
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08b_DEA.html" class="md-nav__link">
        DESeq2
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08c_hypothesis_testing.html" class="md-nav__link">
        Hypothesis testing
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08d_DEA_visualization.html" class="md-nav__link">
        Visualization
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_9" type="checkbox" id="__nav_9" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_9">
          Functional Analysis
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Functional Analysis" data-md-level="1">
        <label class="md-nav__title" for="__nav_9">
          <span class="md-nav__icon md-icon"></span>
          Functional Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="09a_FA_genomic_annotation.html" class="md-nav__link">
        Gene annotation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="09b_FA_overrepresentation.html" class="md-nav__link">
        Over represetation analysis
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Class scoring
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="09c_FA_GSEA.html" class="md-nav__link md-nav__link--active">
        Class scoring
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    Learning Objectives:
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="10_summarized_workflow.html" class="md-nav__link">
        Summarized workflow
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    Learning Objectives:
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                


<h1 id="functional-class-scoring">Functional Class Scoring</h1>
<p>Approximate time: 40 minutes</p>
<h2 id="learning-objectives">Learning Objectives:</h2>
<ul>
<li>Discuss functional class scoring, and pathway topology methods</li>
<li>Construct a GSEA analysis using GO and KEGG gene sets</li>
<li>Examine results of a GSEA using pathview package</li>
<li>List other tools and resources for identifying genes of novel pathways or networks</li>
</ul>
<h1 id="functional-analysis">Functional analysis</h1>
<p>Over-representation analysis is only a single type of functional analysis method that is available for teasing apart the biological processes important to your condition of interest. Other types of analyses can be equally important or informative, including functional class scoring methods. </p>
<p align="center"> 
<img src="./img/09c_FA_GSEA/pathway_analysis.png" width="600">
</p>

<h2 id="functional-class-scoring_1">Functional class scoring</h2>
<p>Functional class scoring (FCS) tools, such as <a href="https://www.pnas.org/content/102/43/15545">GSEA</a>, most often use the gene-level statistics or log2 fold changes for all genes from the differential expression results, then look to see whether gene sets for particular biological pathways are enriched among the large positive or negative fold changes. </p>
<p align="center"> 
<img src="./img/09c_FA_GSEA/gsea_theory.png" width="600">
</p>

<p>The hypothesis of FCS methods is that although large changes in individual genes can have significant effects on pathways (and will be detected via ORA methods), weaker but coordinated changes in sets of functionally related genes (i.e., pathways) can also have significant effects.  Thus, rather than setting an arbitrary threshold to identify 'significant genes', <strong>all genes are considered</strong> in the analysis. The gene-level statistics from the dataset are aggregated to generate a single pathway-level statistic and statistical significance of each pathway is reported. This type of analysis can be particularly helpful if the differential expression analysis only outputs a small list of significant DE genes. </p>
<h3 id="gene-set-enrichment-analysis-using-clusterprofiler-and-pathview">Gene set enrichment analysis using clusterProfiler and Pathview</h3>
<p>Using the log2 fold changes obtained from the differential expression analysis for every gene, gene set enrichment analysis and pathway analysis can be performed using clusterProfiler and Pathview tools.</p>
<p>For a gene set or pathway analysis using clusterProfiler, coordinated differential expression over gene sets is tested instead of changes of individual genes. "Gene sets are pre-defined groups of genes, which are functionally related. Commonly used gene sets include those derived from KEGG pathways, Gene Ontology terms, MSigDB, Reactome, or gene groups that share some other functional annotations, etc. Consistent perturbations over such gene sets frequently suggest mechanistic changes". </p>
<h4 id="preparation-for-gsea">Preparation for GSEA</h4>
<p>clusterProfiler offers several functions to perform GSEA using different genes sets, including but not limited to GO, KEGG, and MSigDb. We will use the KEGG gene sets, which identify genes using their Entrez IDs. Therefore, to perform the analysis, we will need to acquire the Entrez IDs. We will also need to remove the Entrez ID NA values and duplicates (due to gene ID conversion) prior to the analysis:</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Remove any NA values (reduces the data by quite a bit)</span>
<span class="n">res_entrez</span> <span class="o">&lt;-</span> <span class="n">dplyr</span><span class="o">::</span><span class="nf">filter</span><span class="p">(</span><span class="n">res_ids</span><span class="p">,</span> <span class="n">entrezid</span> <span class="o">!=</span> <span class="s">&quot;NA&quot;</span><span class="p">)</span>

<span class="c1">## Remove any Entrez duplicates</span>
<span class="n">res_entrez</span> <span class="o">&lt;-</span> <span class="n">res_entrez</span><span class="p">[</span><span class="nf">which</span><span class="p">(</span><span class="nf">duplicated</span><span class="p">(</span><span class="n">res_entrez</span><span class="o">$</span><span class="n">entrezid</span><span class="p">)</span> <span class="o">==</span> <span class="bp">F</span><span class="p">),</span> <span class="p">]</span>
</code></pre></div>
<p>Finally, extract and name the fold changes:</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Extract the foldchanges</span>
<span class="n">foldchanges</span> <span class="o">&lt;-</span> <span class="n">res_entrez</span><span class="o">$</span><span class="n">log2FoldChange</span>

<span class="c1">## Name each fold change with the corresponding Entrez ID</span>
<span class="nf">names</span><span class="p">(</span><span class="n">foldchanges</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">res_entrez</span><span class="o">$</span><span class="n">entrezid</span>
</code></pre></div>
<p>Next we need to order the fold changes in decreasing order. To do this we'll use the <code>sort()</code> function, which takes a vector as input. This is in contrast to Tidyverse's <code>arrange()</code>, which requires a data frame.</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Sort fold changes in decreasing order</span>
<span class="n">foldchanges</span> <span class="o">&lt;-</span> <span class="nf">sort</span><span class="p">(</span><span class="n">foldchanges</span><span class="p">,</span> <span class="n">decreasing</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>

<span class="nf">head</span><span class="p">(</span><span class="n">foldchanges</span><span class="p">)</span>
</code></pre></div>
<h4 id="theory-of-gsea">Theory of GSEA</h4>
<p>Now we are ready to perform GSEA. The details regarding GSEA can be found in the <a href="https://www.pnas.org/content/102/43/15545">PNAS paper</a> by Subramanian et al. We will describe briefly the steps outlined in the paper below:</p>
<p align="center"> 
<img src="./img/09c_FA_GSEA/gsea_overview.png" width="600">
</p>

<p><em><strong>Image credit:</strong> <a href="https://www.pnas.org/content/102/43/15545">Subramanian et al. Proceedings of the National Academy of Sciences Oct 2005, 102 (43) 15545-15550; DOI: 10.1073/pnas.0506580102</a></em></p>
<p>This image describes the theory of GSEA, with the 'gene set S' showing the metric used (in our case, ranked log2 fold changes) to determine enrichment of genes in the gene set. The left-most image is representing this metric used for the GSEA analysis. The log2 fold changes for each gene in the 'gene set S' is shown as a line in the middle image. The large positive log2 fold changes are at the top of the gene set image, while the largest negative log2 fold changes are at the bottom of the gene set image. In the right-most image, the gene set is turned horizontally, underneath which is an image depicting the calculations involved in determining enrichment, as described below.</p>
<p><strong>Step 1:</strong> Calculation of enrichment score: </p>
<p>An enrichment score for a particular gene set is calculated by walking down the list of log2 fold changes and increasing the running-sum statistic every time a gene in the gene set is encountered and decreasing it when genes are not part of the gene set. The size of the increase/decrease is determined by magnitude of the log2 fold change. Larger (positive or negative) log2 fold changes will result in larger increases or decreases. The final enrichment score is where the running-sum statistic is the largest deviation from zero.</p>
<p><strong>Step 2:</strong> Estimation of significance:</p>
<p>The significance of the enrichment score is determined using permutation testing, which performs rearrangements of the data points to determine the likelihood of generating an enrichment score as large as the enrichment score calculated from the observed data. Essentially, for this step, the first permutation would reorder the log2 fold changes and randomly assign them to different genes, reorder the gene ranks based on these new log2 fold changes, and recalculate the enrichment score. The second permutation would reorder the log2 fold changes again and recalculate the enrichment score again, and this would continue for the total number of permutations run. Therefore, the number of permutations run will increase the confidence in the signficance estimates.</p>
<p><strong>Step 3:</strong> Adjust for multiple test correction</p>
<p>After all gene sets are tested, the enrichment scores are normalized for the size of the gene set, then the p-values are corrected for multiple testing.</p>
<p>The GSEA output will yield the core genes in the gene sets that most highly contribute to the enrichment score. The genes output are generally the genes at or before the running sum reaches its maximum value (eg. the most influential genes driving the differences between conditions for that gene set).</p>
<h4 id="performing-gsea">Performing GSEA</h4>
<p>To perform the GSEA using KEGG gene sets with clusterProfiler, we can use the <code>gseKEGG()</code> function:</p>
<div class="highlight"><pre><span></span><code><span class="c1">## GSEA using gene sets from KEGG pathways</span>
<span class="n">gseaKEGG</span> <span class="o">&lt;-</span> <span class="nf">gseKEGG</span><span class="p">(</span><span class="n">geneList</span> <span class="o">=</span> <span class="n">foldchanges</span><span class="p">,</span> <span class="c1"># ordered named vector of fold changes (Entrez IDs are the associated names)</span>
              <span class="n">organism</span> <span class="o">=</span> <span class="s">&quot;hsa&quot;</span><span class="p">,</span> <span class="c1"># supported organisms listed below</span>
              <span class="n">nPerm</span> <span class="o">=</span> <span class="m">1000</span><span class="p">,</span> <span class="c1"># default number permutations</span>
              <span class="n">minGSSize</span> <span class="o">=</span> <span class="m">20</span><span class="p">,</span> <span class="c1"># minimum gene set size (# genes in set) - change to test more sets or recover sets with fewer # genes</span>
              <span class="n">pvalueCutoff</span> <span class="o">=</span> <span class="m">0.05</span><span class="p">,</span> <span class="c1"># padj cutoff value</span>
              <span class="n">verbose</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>

<span class="c1">## Extract the GSEA results</span>
<span class="n">gseaKEGG_results</span> <span class="o">&lt;-</span> <span class="n">gseaKEGG</span><span class="o">@</span><span class="n">result</span>
</code></pre></div>
<blockquote>
<p><strong>NOTE:</strong> The organisms with KEGG pathway information are listed <a href="http://www.genome.jp/kegg/catalog/org_list.html">here</a>.</p>
</blockquote>
<p><strong>How many pathways are enriched?</strong> View the enriched pathways:</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Write GSEA results to file</span>
<span class="nf">View</span><span class="p">(</span><span class="n">gseaKEGG_results</span><span class="p">)</span>

<span class="nf">write.csv</span><span class="p">(</span><span class="n">gseaKEGG_results</span><span class="p">,</span> <span class="s">&quot;results/gseaOE_kegg.csv&quot;</span><span class="p">,</span> <span class="n">quote</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
</code></pre></div>
<blockquote>
<p><em><strong>NOTE:</strong> We will all get different results for the GSEA because the permutations performed use random reordering. If we would like to use the same permutations every time we run a function (i.e. we would like the same results every time we run the function), then we could use the <code>set.seed(123456)</code> function prior to running. The input to <code>set.seed()</code> could be any number, but if you would want the same results, then you would need to use the same number as input.</em></p>
</blockquote>
<p>Explore the GSEA plot of enrichment of one of the pathways in the ranked list:</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Plot the GSEA plot for a single enriched pathway, `hsa03040`</span>
<span class="nf">gseaplot</span><span class="p">(</span><span class="n">gseaKEGG</span><span class="p">,</span> <span class="n">geneSetID</span> <span class="o">=</span> <span class="s">&#39;hsa03040&#39;</span><span class="p">)</span>
</code></pre></div>
<p align="center"> 
<img src="./img/09c_FA_GSEA/gsea_kegg_hsa03040.png" width="600">
</p>

<p>In this plot, the lines in plot represent the genes in the gene set 'hsa03040', and where they occur among the log2 fold changes. The largest positive log2 fold changes are on the left-hand side of the plot, while the largest negative log2 fold changes are on the right. The top plot shows the magnitude of the log2 fold changes for each gene, while the bottom plot shows the running sum, with the enrichment score peaking at the red dotted line (which is among the negative log2 fold changes).</p>
<p>Use the <a href="http://bioconductor.org/packages/release/bioc/html/pathview.html">Pathview R package</a> to integrate the KEGG pathway data from clusterProfiler into pathway images:</p>
<div class="highlight"><pre><span></span><code><span class="nf">detach</span><span class="p">(</span><span class="s">&quot;package:dplyr&quot;</span><span class="p">,</span> <span class="n">unload</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span> <span class="c1"># first unload dplyr to avoid conflicts</span>

<span class="c1">## Output images for a single significant KEGG pathway</span>
<span class="nf">pathview</span><span class="p">(</span><span class="n">gene.data</span> <span class="o">=</span> <span class="n">foldchanges</span><span class="p">,</span>
              <span class="n">pathway.id</span> <span class="o">=</span> <span class="s">&quot;hsa03040&quot;</span><span class="p">,</span>
              <span class="n">species</span> <span class="o">=</span> <span class="s">&quot;hsa&quot;</span><span class="p">,</span>
              <span class="n">limit</span> <span class="o">=</span> <span class="nf">list</span><span class="p">(</span><span class="n">gene</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> <span class="c1"># value gives the max/min limit for foldchanges</span>
              <span class="n">cpd</span> <span class="o">=</span> <span class="m">1</span><span class="p">))</span>
</code></pre></div>
<blockquote>
<p><strong>NOTE:</strong> If the below error message occurs: <code>Error in detach("package:dplyr", unload = T) : invalid 'name' argument</code>, that means the dplyr package is not currently loaded. Ignore the message and continue to run pathview command.</p>
</blockquote>
<p align="center"> 
<img src="./img/09c_FA_GSEA/hsa03040.pathview.png" width="800">
</p>

<blockquote>
<p><strong>NOTE:</strong> Printing out Pathview images for all significant pathways can be easily performed as follows:</p>
<p>```r</p>
<h2 id="output-images-for-all-significant-kegg-pathways">Output images for all significant KEGG pathways</h2>
<p>get_kegg_plots &lt;- function(x) {
   pathview(gene.data = foldchanges, 
            pathway.id = gseaKEGG_results$ID[x], 
            species = "hsa",
            limit = list(gene = 2, cpd = 1))
}</p>
<p>purrr::map(1:length(gseaKEGG_results$ID), 
           get_kegg_plots)
```</p>
</blockquote>
<p>Instead of exploring enrichment of KEGG gene sets, we can also explore the enrichment of BP Gene Ontology terms using gene set enrichment analysis: </p>
<div class="highlight"><pre><span></span><code><span class="c1"># GSEA using gene sets associated with BP Gene Ontology terms</span>
<span class="n">gseaGO</span> <span class="o">&lt;-</span> <span class="nf">gseGO</span><span class="p">(</span><span class="n">geneList</span> <span class="o">=</span> <span class="n">foldchanges</span><span class="p">,</span> 
              <span class="n">OrgDb</span> <span class="o">=</span> <span class="n">org.Hs.eg.db</span><span class="p">,</span> 
              <span class="n">ont</span> <span class="o">=</span> <span class="s">&#39;BP&#39;</span><span class="p">,</span> 
              <span class="n">nPerm</span> <span class="o">=</span> <span class="m">1000</span><span class="p">,</span> 
              <span class="n">minGSSize</span> <span class="o">=</span> <span class="m">20</span><span class="p">,</span> 
              <span class="n">pvalueCutoff</span> <span class="o">=</span> <span class="m">0.05</span><span class="p">,</span>
              <span class="n">verbose</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span> 

<span class="n">gseaGO_results</span> <span class="o">&lt;-</span> <span class="n">gseaGO</span><span class="o">@</span><span class="n">result</span>

<span class="nf">gseaplot</span><span class="p">(</span><span class="n">gseaGO</span><span class="p">,</span> <span class="n">geneSetID</span> <span class="o">=</span> <span class="s">&#39;GO:0007423&#39;</span><span class="p">)</span>
</code></pre></div>
<p>There are other gene sets available for GSEA analysis in clusterProfiler (Disease Ontology, Reactome pathways, etc.). In addition, it is possible to supply your own gene set GMT file, such as a GMT for <a href="http://software.broadinstitute.org/gsea/msigdb/index.jsp">MSigDB</a> using special clusterProfiler functions as shown below:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># DO NOT RUN</span>

<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;GSEABase&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">GSEABase</span><span class="p">)</span>

<span class="c1"># Load in GMT file of gene sets (we downloaded from the Broad Institute for MSigDB)</span>

<span class="n">c2</span> <span class="o">&lt;-</span> <span class="nf">read.gmt</span><span class="p">(</span><span class="s">&quot;/data/c2.cp.v6.0.entrez.gmt.txt&quot;</span><span class="p">)</span>

<span class="n">msig</span> <span class="o">&lt;-</span> <span class="nf">GSEA</span><span class="p">(</span><span class="n">foldchanges</span><span class="p">,</span> <span class="n">TERM2GENE</span><span class="o">=</span><span class="n">c2</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>

<span class="n">msig_df</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">msig</span><span class="p">)</span>
</code></pre></div>
<h2 id="pathway-topology-tools">Pathway topology tools</h2>
<p align="center"> 
<img src="./img/09c_FA_GSEA/pathway_analysis.png" width="600">
</p>

<p>The last main type of functional analysis technique is pathway topology analysis. Pathway topology analysis often takes into account gene interaction information along with the fold changes and adjusted p-values from differential expression analysis to identify dysregulated pathways. Depending on the tool, pathway topology tools explore how genes interact with each other (e.g. activation, inhibition, phosphorylation, ubiquitination, etc.) to determine the pathway-level statistics. Pathway topology-based methods utilize the number and type of interactions between gene product (our DE genes) and other gene products to infer gene function or pathway association. </p>
<p>For instance, the <a href="http://bioconductor.org/packages/release/bioc/html/SPIA.html">SPIA (Signaling Pathway Impact Analysis)</a> tool can be used to integrate the lists of differentially expressed genes, their fold changes, and pathway topology to identify affected pathways.</p>
<h2 id="other-tools">Other Tools</h2>
<h3 id="co-expression-clustering">Co-expression clustering</h3>
<p>Co-expression clustering is often used to identify genes of novel pathways or networks by grouping genes together based on similar trends in expression. These tools are useful in identifying genes in a pathway, when their participation in a pathway and/or the pathway itself is unknown. These tools cluster genes with similar expression patterns to create 'modules' of co-expressed genes which often reflect functionally similar groups of genes. These 'modules' can then be compared across conditions or in a time-course experiment to identify any biologically relevant pathway or network information.</p>
<p>You can visualize co-expression clustering using heatmaps, which should be viewed as suggestive only; serious classification of genes needs better methods.  </p>
<p>The way the tools perform clustering is by taking the entire expression matrix and computing pair-wise co-expression values. A network is then generated from which we explore the topology to make inferences on gene co-regulation. The <a href="https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/">WGCNA</a> package (in R) is one example of a more sophisticated method for co-expression clustering.</p>
<h2 id="resources-for-functional-analysis">Resources for functional analysis</h2>
<ul>
<li>g:Profiler - http://biit.cs.ut.ee/gprofiler/index.cgi </li>
<li>DAVID - http://david.abcc.ncifcrf.gov/tools.jsp </li>
<li>clusterProfiler - http://bioconductor.org/packages/release/bioc/html/clusterProfiler.html</li>
<li>GeneMANIA - http://www.genemania.org/</li>
<li>GenePattern -  http://www.broadinstitute.org/cancer/software/genepattern/ (need to register)</li>
<li>WebGestalt - http://bioinfo.vanderbilt.edu/webgestalt/ (need to register)</li>
<li>AmiGO - http://amigo.geneontology.org/amigo</li>
<li>ReviGO (visualizing GO analysis, input is GO terms) - http://revigo.irb.hr/ </li>
<li>WGCNA - https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/</li>
<li>GSEA - http://software.broadinstitute.org/gsea/index.jsp</li>
<li>SPIA - https://www.bioconductor.org/packages/release/bioc/html/SPIA.html</li>
<li>GAGE/Pathview - http://www.bioconductor.org/packages/release/bioc/html/gage.html</li>
</ul>
<hr />
<p><em>This lesson was originally developed by members of the teaching team (Mary Piper, Radhika Khetani) at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>.</em></p>

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="09b_FA_overrepresentation.html" class="md-footer__link md-footer__link--prev" aria-label="Previous: Over represetation analysis" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Over represetation analysis
            </div>
          </div>
        </a>
      
      
        
        <a href="10_summarized_workflow.html" class="md-footer__link md-footer__link--next" aria-label="Next: Summarized workflow" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Summarized workflow
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": ".", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "assets/javascripts/workers/search.bd0b6b67.min.js"}</script>
    
    
      <script src="assets/javascripts/bundle.467223ff.min.js"></script>
      
    
  </body>
</html>