
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="img/sandbox1.png">
      <meta name="generator" content="mkdocs-1.3.0, mkdocs-material-8.2.14">
    
    
      
        <title>Summarized workflow - Bulk RNAseq data analysis</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.3de6f41f.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.cc9b2e1e.min.css">
        
          
          
          <meta name="theme-color" content="#757575">
        
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="extra.css">
    
    <script>__md_scope=new URL(".",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="grey" data-md-color-accent="cyan">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-objectives" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="https://hds-sandbox.github.io/" title="Bulk RNAseq data analysis" class="md-header__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component="logo">
      
  <img src="img/sandbox1.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bulk RNAseq data analysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Summarized workflow
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="https://hds-sandbox.github.io/" title="Bulk RNAseq data analysis" class="md-nav__button md-logo" aria-label="Bulk RNAseq data analysis" data-md-component="logo">
      
  <img src="img/sandbox1.png" alt="logo">

    </a>
    Bulk RNAseq data analysis
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="index.html" class="md-nav__link">
        Course Introduction
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="02_setup.html" class="md-nav__link">
        Setup
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="03_experimental_planning.html" class="md-nav__link">
        Experimental planning
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="04_data_explanation.html" class="md-nav__link">
        Data Explanation
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="05_preprocessing.html" class="md-nav__link">
        Preprocessing
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          RNAseq counts
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="RNAseq counts" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          RNAseq counts
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="06a_count_matrix.html" class="md-nav__link">
        Count matrix
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="06b_count_normalization.html" class="md-nav__link">
        Normalization
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="07_exploratory_analysis.html" class="md-nav__link">
        Exploratory Analysis
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_8" type="checkbox" id="__nav_8" >
      
      
      
      
        <label class="md-nav__link" for="__nav_8">
          Differential Expression
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Differential Expression" data-md-level="1">
        <label class="md-nav__title" for="__nav_8">
          <span class="md-nav__icon md-icon"></span>
          Differential Expression
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08a_contrast_design.html" class="md-nav__link">
        Design formula
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08b_DEA.html" class="md-nav__link">
        DESeq2
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08c_hypothesis_testing.html" class="md-nav__link">
        Hypothesis testing
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="08d_DEA_visualization.html" class="md-nav__link">
        Visualization
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_9" type="checkbox" id="__nav_9" >
      
      
      
      
        <label class="md-nav__link" for="__nav_9">
          Functional Analysis
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Functional Analysis" data-md-level="1">
        <label class="md-nav__title" for="__nav_9">
          <span class="md-nav__icon md-icon"></span>
          Functional Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="09a_FA_genomic_annotation.html" class="md-nav__link">
        Gene annotation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="09b_FA_overrepresentation.html" class="md-nav__link">
        Over represetation analysis
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="09c_FA_GSEA.html" class="md-nav__link">
        Class scoring
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Summarized workflow
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="10_summarized_workflow.html" class="md-nav__link md-nav__link--active">
        Summarized workflow
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    Learning Objectives
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summary-of-differential-expression-analysis-workflow" class="md-nav__link">
    Summary of differential expression analysis workflow
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    Learning Objectives
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summary-of-differential-expression-analysis-workflow" class="md-nav__link">
    Summary of differential expression analysis workflow
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                


  <h1>Summarized workflow</h1>

<p>Approximate time: 15 minutes</p>
<h2 id="learning-objectives">Learning Objectives</h2>
<ul>
<li>Identify the R commands needed to run a complete differential
    expression analysis using DESeq2</li>
</ul>
<h2 id="summary-of-differential-expression-analysis-workflow">Summary of differential expression analysis workflow</h2>
<p>We have detailed the various steps in a differential expression analysis
workflow, providing theory with example code. To provide a more succinct
reference for the code needed to run a DGE analysis, we have summarized
the steps in an analysis below:</p>
<ol>
<li>Obtaining gene-level counts from your preprocessing</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="n">data</span> <span class="o">&lt;-</span> <span class="nf">read_table</span><span class="p">(</span><span class="s">&quot;data/Mov10_full_counts.txt&quot;</span><span class="p">)</span> 

<span class="n">meta</span> <span class="o">&lt;-</span> <span class="nf">read_table</span><span class="p">(</span><span class="s">&quot;meta/Mov10_full_meta.txt&quot;</span><span class="p">)</span>
</code></pre></div>
<ol>
<li>Creating the dds object:</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="c1"># Check that the row names of the metadata equal the column names of the **raw counts** data</span>
<span class="nf">all</span><span class="p">(</span><span class="nf">colnames</span><span class="p">(</span><span class="n">datacounts</span><span class="p">)[</span><span class="m">-1</span><span class="p">]</span> <span class="o">==</span> <span class="n">meta</span><span class="o">$</span><span class="n">samplename</span><span class="p">)</span>

<span class="c1"># Create DESeq2Dataset object</span>
<span class="n">dds</span> <span class="o">&lt;-</span> <span class="nf">DESeqDataSetFromMatrix</span><span class="p">(</span><span class="n">countData</span> <span class="o">=</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">data</span><span class="p">[,</span><span class="m">-1</span><span class="p">],</span> <span class="n">row.names</span> <span class="o">=</span> <span class="n">data</span><span class="o">$</span><span class="n">GeneSymbol</span><span class="p">),</span> 
                              <span class="n">colData</span> <span class="o">=</span> <span class="n">meta</span><span class="p">,</span> 
                              <span class="n">design</span> <span class="o">=</span> <span class="o">~</span> <span class="n">sampletype</span><span class="p">)</span>
</code></pre></div>
<ol>
<li>Exploratory data analysis (PCA &amp; hierarchical clustering) -
    identifying outliers and sources of variation in the data:</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="c1"># Transform counts for data visualization</span>
<span class="n">rld</span> <span class="o">&lt;-</span> <span class="nf">rlog</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span> 
            <span class="n">blind</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>

<span class="c1"># Plot PCA </span>
<span class="nf">plotPCA</span><span class="p">(</span><span class="n">rld</span><span class="p">,</span> 
        <span class="n">intgroup</span><span class="o">=</span><span class="s">&quot;sampletype&quot;</span><span class="p">)</span>

<span class="c1"># Extract the rlog matrix from the object and compute pairwise correlation values</span>
<span class="n">rld_mat</span> <span class="o">&lt;-</span> <span class="nf">assay</span><span class="p">(</span><span class="n">rld</span><span class="p">)</span>
<span class="n">rld_cor</span> <span class="o">&lt;-</span> <span class="nf">cor</span><span class="p">(</span><span class="n">rld_mat</span><span class="p">)</span>

<span class="c1"># Plot heatmap</span>
<span class="nf">pheatmap</span><span class="p">(</span><span class="n">rld_cor</span><span class="p">,</span> 
         <span class="n">annotation</span> <span class="o">=</span> <span class="n">meta</span> <span class="o">%&gt;%</span> <span class="nf">column_to_rownames</span><span class="p">(</span><span class="s">&quot;samplename&quot;</span><span class="p">))</span>
</code></pre></div>
<ol>
<li>Run DESeq2:</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="c1"># **Optional step** - Re-create DESeq2 dataset if the design formula has changed after QC analysis in include other sources of variation using &quot;dds &lt;- DESeqDataSetFromMatrix(data, colData = metadata, design = ~ covariate + condition)&quot;</span>

<span class="c1"># Run DESeq2 differential expression analysis</span>
<span class="n">dds</span> <span class="o">&lt;-</span> <span class="nf">DESeq</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>

<span class="c1"># **Optional step** - Output normalized counts to save as a file to access outside RStudio using &quot;normalized_counts &lt;- counts(dds, normalized=TRUE)&quot;</span>
</code></pre></div>
<ol>
<li>Check the fit of the dispersion estimates:</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="c1"># Plot dispersion estimates</span>
<span class="nf">plotDispEsts</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
</code></pre></div>
<ol>
<li>Create contrasts to perform Wald testing on the shrunken log2
    foldchanges between specific conditions:</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="c1"># Specify contrast for comparison of interest</span>
<span class="n">contrast</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;condition&quot;</span><span class="p">,</span> <span class="s">&quot;level_to_compare&quot;</span><span class="p">,</span> <span class="s">&quot;base_level&quot;</span><span class="p">)</span>

<span class="c1"># Output results of Wald test for contrast</span>
<span class="n">res</span> <span class="o">&lt;-</span> <span class="nf">results</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span> 
               <span class="n">contrast</span> <span class="o">=</span> <span class="n">contrast</span><span class="p">,</span> 
               <span class="n">alpha</span> <span class="o">=</span> <span class="m">0.05</span><span class="p">)</span>

<span class="c1"># Shrink the log2 fold changes to be more accurate</span>
<span class="n">res</span> <span class="o">&lt;-</span> <span class="nf">lfcShrink</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span> 
                 <span class="n">contrast</span> <span class="o">=</span> <span class="n">contrast</span><span class="p">,</span> 
                 <span class="n">type</span> <span class="o">=</span> <span class="s">&quot;normal&quot;</span><span class="p">)</span>
</code></pre></div>
<ol>
<li>Output significant results:</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="c1"># Set thresholds</span>
<span class="n">padj.cutoff</span> <span class="o">&lt;</span> <span class="o">-</span> <span class="m">0.05</span>

<span class="c1"># Turn the results object into a tibble for use with tidyverse functions</span>
<span class="n">res_tbl</span> <span class="o">&lt;-</span> <span class="n">res</span> <span class="o">%&gt;%</span>
  <span class="nf">data.frame</span><span class="p">()</span> <span class="o">%&gt;%</span>
  <span class="nf">rownames_to_column</span><span class="p">(</span><span class="n">var</span><span class="o">=</span><span class="s">&quot;gene&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span> 
  <span class="nf">as_tibble</span><span class="p">()</span>

<span class="c1"># Subset the significant results</span>
<span class="n">sig_res</span> <span class="o">&lt;-</span> <span class="nf">filter</span><span class="p">(</span><span class="n">res_tbl</span><span class="p">,</span> 
                  <span class="n">padj</span> <span class="o">&lt;</span> <span class="n">padj.cutoff</span><span class="p">)</span>
</code></pre></div>
<ol>
<li>
<p>Visualize results: volcano plots, heatmaps, normalized counts plots
    of top genes, etc.</p>
</li>
<li>
<p>Perform analysis to extract functional significance of results: GO
    or KEGG enrichment, GSEA, etc.</p>
</li>
<li>
<p>Make sure to output the versions of all tools used in the DE
    analysis:</p>
</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="nf">sessionInfo</span><span class="p">()</span>
</code></pre></div>
<p>For better reproducibility, it can help to create <strong>RMarkdown reports</strong>,
which save all code, results, and visualizations as nicely formatted
html reports. We have available an <a href="https://www.dropbox.com/s/xp7l0b46rvwmbya/workshop-example.html?dl=1">example html
report</a>
for perusal. To create these reports we have <a href="https://hbctraining.github.io/Training-modules/Rmarkdown/">additional
materials</a>
available.</p>
<hr />
<p><em>This lesson was originally developed by members of the teaching team
(Mary Piper) at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core
(HBC)</a>.</em></p>

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="09c_FA_GSEA.html" class="md-footer__link md-footer__link--prev" aria-label="Previous: Class scoring" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Class scoring
            </div>
          </div>
        </a>
      
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": ".", "features": [], "search": "assets/javascripts/workers/search.2a1c317c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="assets/javascripts/bundle.c2e1ee47.min.js"></script>
      
    
  </body>
</html>